// @project-chip/matter.js/dist/esm/cluster/definitions/DeviceEnergyManagementModeCluster.js
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MutableCluster } from "../mutation/MutableCluster.js";
import {
  WritableAttribute,
  FixedAttribute,
  Attribute,
  OptionalWritableAttribute,
  Command,
  TlvNoResponse
} from "../Cluster.js";
import { TlvUInt8 } from "../../tlv/TlvNumber.js";
import { TlvNullable } from "../../tlv/TlvNullable.js";
import { BitFlag } from "../../schema/BitmapSchema.js";
import { TlvArray } from "../../tlv/TlvArray.js";
import { ModeBase } from "./ModeBaseCluster.js";
import { ClusterRegistry } from "../ClusterRegistry.js";
var DeviceEnergyManagementMode;
((DeviceEnergyManagementMode2) => {
  let Feature;
  ((Feature2) => {
    Feature2["OnOff"] = "OnOff";
  })(Feature = DeviceEnergyManagementMode2.Feature || (DeviceEnergyManagementMode2.Feature = {}));
  let ModeTag;
  ((ModeTag2) => {
    ModeTag2[ModeTag2["NoOptimization"] = 16384] = "NoOptimization";
    ModeTag2[ModeTag2["DeviceOptimization"] = 16385] = "DeviceOptimization";
    ModeTag2[ModeTag2["LocalOptimization"] = 16386] = "LocalOptimization";
    ModeTag2[ModeTag2["GridOptimization"] = 16387] = "GridOptimization";
  })(ModeTag = DeviceEnergyManagementMode2.ModeTag || (DeviceEnergyManagementMode2.ModeTag = {}));
  DeviceEnergyManagementMode2.OnOffComponent = MutableCluster.Component({
    attributes: {
      /**
       * Indicates whether the value of CurrentMode depends on the state of the On/Off cluster on the same
       * endpoint. If this attribute is not present or is set to null, there is no dependency, otherwise the
       * CurrentMode attribute shall depend on the OnOff attribute in the On/Off cluster
       *
       * The value of this field shall match the Mode field of one of the entries in the SupportedModes attribute.
       *
       * @see {@link MatterSpecification.v13.Cluster} § 1.10.6.5
       */
      onMode: WritableAttribute(3, TlvNullable(TlvUInt8), { persistent: true, default: null })
    }
  });
  DeviceEnergyManagementMode2.Base = MutableCluster.Component({
    id: 159,
    name: "DeviceEnergyManagementMode",
    revision: 1,
    features: {
      /**
       * OnOff
       *
       * This feature creates a dependency between an OnOff cluster instance and this cluster instance on the
       * same endpoint. See OnMode for more information.
       *
       * @see {@link MatterSpecification.v13.Cluster} § 1.10.4.1
       */
      onOff: BitFlag(0)
    },
    attributes: {
      /**
       * This attribute shall contain the list of supported modes that may be selected for the CurrentMode
       * attribute. Each item in this list represents a unique mode as indicated by the Mode field of the
       * ModeOptionStruct.
       *
       * Each entry in this list shall have a unique value for the Mode field. Each entry in this list shall have
       * a unique value for the Label field.
       *
       * @see {@link MatterSpecification.v13.Cluster} § 1.10.6.2
       */
      supportedModes: FixedAttribute(0, TlvArray(ModeBase.TlvModeOption, { minLength: 2, maxLength: 255 })),
      /**
       * Indicates the current mode of the server.
       *
       * The value of this field shall match the Mode field of one of the entries in the SupportedModes attribute.
       *
       * The value of this attribute may change at any time via an out-of-band interaction outside of the server,
       * such as interactions with a user interface, via internal mode changes due to autonomously progressing
       * through a sequence of operations, on system time-outs or idle delays, or via interactions coming from a
       * fabric other than the one which last executed a ChangeToMode.
       *
       * @see {@link MatterSpecification.v13.Cluster} § 1.10.6.3
       */
      currentMode: Attribute(1, TlvUInt8, { scene: true, persistent: true }),
      /**
       * Indicates the desired startup mode for the server when it is supplied with power.
       *
       * If this attribute is not null, the CurrentMode attribute shall be set to the StartUpMode value, when the
       * server is powered up, except in the case when the OnMode attribute overrides the StartUpMode attribute
       * (see OnModeWithPowerUp).
       *
       * This behavior does not apply to reboots associated with OTA. After an OTA restart, the CurrentMode
       * attribute shall return to its value prior to the restart.
       *
       * The value of this field shall match the Mode field of one of the entries in the SupportedModes attribute.
       *
       * If this attribute is not implemented, or is set to the null value, it shall have no effect.
       *
       * @see {@link MatterSpecification.v13.Cluster} § 1.10.6.4
       */
      startUpMode: OptionalWritableAttribute(2, TlvNullable(TlvUInt8), { persistent: true })
    },
    commands: {
      /**
       * This command is used to change device modes.
       *
       * On receipt of this command the device shall respond with a ChangeToModeResponse command.
       *
       * @see {@link MatterSpecification.v13.Cluster} § 1.10.7.1
       */
      changeToMode: Command(0, ModeBase.TlvChangeToModeRequest, 0, TlvNoResponse)
    },
    /**
     * This metadata controls which DeviceEnergyManagementModeCluster elements matter.js activates for specific
     * feature combinations.
     */
    extensions: MutableCluster.Extensions({ flags: { onOff: true }, component: DeviceEnergyManagementMode2.OnOffComponent })
  });
  DeviceEnergyManagementMode2.ClusterInstance = MutableCluster(DeviceEnergyManagementMode2.Base);
  DeviceEnergyManagementMode2.Cluster = DeviceEnergyManagementMode2.ClusterInstance;
  const DEPONOFF = { onOff: true };
  DeviceEnergyManagementMode2.CompleteInstance = MutableCluster({
    id: DeviceEnergyManagementMode2.Cluster.id,
    name: DeviceEnergyManagementMode2.Cluster.name,
    revision: DeviceEnergyManagementMode2.Cluster.revision,
    features: DeviceEnergyManagementMode2.Cluster.features,
    attributes: {
      ...DeviceEnergyManagementMode2.Cluster.attributes,
      onMode: MutableCluster.AsConditional(DeviceEnergyManagementMode2.OnOffComponent.attributes.onMode, { mandatoryIf: [DEPONOFF] })
    },
    commands: DeviceEnergyManagementMode2.Cluster.commands
  });
  DeviceEnergyManagementMode2.Complete = DeviceEnergyManagementMode2.CompleteInstance;
})(DeviceEnergyManagementMode || (DeviceEnergyManagementMode = {}));
const DeviceEnergyManagementModeCluster = DeviceEnergyManagementMode.Cluster;
ClusterRegistry.register(DeviceEnergyManagementMode.Complete);
export {
  DeviceEnergyManagementMode,
  DeviceEnergyManagementModeCluster
};
