// @project-chip/matter.js/dist/esm/net/fake/SimulatedNetwork.js
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Logger } from "../../log/Logger.js";
import { singleton } from "../../util/Singleton.js";
const logger = Logger.get("SimulatedNetwork");
const FAKE_INTERFACE_NAME = "fakeInterface";
class SimulatedNetwork {
  static get = singleton(() => new SimulatedNetwork());
  listenersMap = /* @__PURE__ */ new Map();
  onUdpData(host, port, listener) {
    const ipPort = `${host ?? "*"}:${port}`;
    let listeners = this.listenersMap.get(ipPort);
    if (listeners === void 0) {
      listeners = new Array();
      this.listenersMap.set(ipPort, listeners);
    }
    listeners.push(listener);
    return {
      close: async () => this.offUdpData(host, port, listener)
    };
  }
  offUdpData(host, port, listenerToRemove) {
    const ipPort = `${host ?? "*"}:${port}`;
    const listeners = this.listenersMap.get(ipPort);
    if (listeners === void 0) return;
    const newListeners = listeners.filter((listener) => listener !== listenerToRemove);
    if (newListeners.length === 0) {
      this.listenersMap.delete(ipPort);
      return;
    }
    this.listenersMap.set(ipPort, newListeners);
  }
  sendUdp(localAddress, localPort, remoteAddress, remotePort, data) {
    [`${remoteAddress}:${remotePort}`, `*:${remotePort}`].forEach(
      (ipPort) => this.listenersMap.get(ipPort)?.forEach((listener) => {
        try {
          listener(FAKE_INTERFACE_NAME, localAddress, localPort, data);
        } catch (error) {
          logger.error(error);
        }
      })
    );
  }
}
export {
  FAKE_INTERFACE_NAME,
  SimulatedNetwork
};
