// @project-chip/matter.js/dist/cjs/node/server/storage/ServerStore.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ServerStore_exports = {};
__export(ServerStore_exports, {
  ServerStore: () => ServerStore
});
module.exports = __toCommonJS(ServerStore_exports);
var import_MatterError = require("../../../common/MatterError.js");
var import_StorageService = require("../../../environment/StorageService.js");
var import_Diagnostic = require("../../../log/Diagnostic.js");
var import_Logger = require("../../../log/Logger.js");
var import_EventHandler = require("../../../protocol/interaction/EventHandler.js");
var import_Construction = require("../../../util/Construction.js");
var import_PartStoreService = require("./PartStoreService.js");
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const logger = import_Logger.Logger.get("ServerStore");
class ServerStore {
  #location;
  #nodeId;
  #storageManager;
  #eventHandler;
  #sessionStorage;
  #fabricStorage;
  #eventStorage;
  #rootStore;
  #construction;
  get construction() {
    return this.#construction;
  }
  /**
   * Create a new store.
   *
   * TODO - implement conversion from 0.7 format so people can change API seamlessly
   */
  constructor(environment, nodeId) {
    if (nodeId === void 0) {
      throw new import_MatterError.ImplementationError("ServerStore must be created with a nodeId");
    }
    const storage = environment.get(import_StorageService.StorageService);
    this.#location = storage.location ?? "(unknown location)";
    this.#nodeId = nodeId;
    const initializeStorage = async () => {
      this.#storageManager = await storage.open(nodeId);
      this.#rootStore = await (0, import_Construction.asyncNew)(import_PartStoreService.PartStoreFactory, {
        storage: this.#storageManager.createContext("root")
      });
      this.#eventStorage = this.#storage.createContext("events");
      this.#eventHandler = await import_EventHandler.EventHandler.create(this.eventStorage);
      this.#logChange("Opened");
    };
    this.#construction = (0, import_Construction.Construction)(this, initializeStorage);
  }
  static async create(environment, nodeId) {
    return await (0, import_Construction.asyncNew)(this, environment, nodeId);
  }
  async erase() {
    await this.#sessionStorage?.clearAll();
    await this.#fabricStorage?.clearAll();
    await this.#eventStorage?.clearAll();
    await this.#rootStore?.erase();
  }
  async close() {
    await this.#construction.close(async () => {
      await this.#storageManager?.close();
      this.#logChange("Closed");
    });
  }
  get eventStorage() {
    if (!this.#eventStorage) {
      throw new import_MatterError.ImplementationError("Event storage accessed prior to initialization");
    }
    return this.#eventStorage;
  }
  get eventHandler() {
    if (!this.#eventHandler) {
      throw new import_MatterError.ImplementationError("Event handler accessed prior to initialization");
    }
    return this.#eventHandler;
  }
  get sessionStorage() {
    if (!this.#sessionStorage) {
      this.#sessionStorage = this.#storage.createContext("sessions");
    }
    return this.#sessionStorage;
  }
  get fabricStorage() {
    if (!this.#fabricStorage) {
      this.#fabricStorage = this.#storage.createContext("fabrics");
    }
    return this.#fabricStorage;
  }
  get partStores() {
    if (this.#rootStore === void 0) {
      throw new import_MatterError.ImplementationError("Endpoint storage accessed prior to initialization");
    }
    return this.#rootStore;
  }
  get #storage() {
    if (this.#storageManager === void 0) {
      throw new import_MatterError.ImplementationError("Node storage accessed prior to initialization");
    }
    return this.#storageManager;
  }
  #logChange(what) {
    logger.info(what, import_Diagnostic.Diagnostic.strong(this.#nodeId ?? "node"), "storage at", `${this.#location}/${this.#nodeId}`);
  }
}
