// @project-chip/matter.js/dist/cjs/util/CommandlineParser.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var CommandlineParser_exports = {};
__export(CommandlineParser_exports, {
  commandlineParser: () => commandlineParser
});
module.exports = __toCommonJS(CommandlineParser_exports);
/**
 * Quote aware Command line parser
 *
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class StringReader {
  constructor(str) {
    this.str = str;
    this.pos = -1;
  }
  pos;
  current;
  next() {
    if (this.pos < this.str.length) {
      this.current = this.str[++this.pos];
    }
    return this.current;
  }
}
function isWhitespace(c) {
  return c === " ";
}
function consumeWhitespace(r) {
  while (r.current !== void 0 && isWhitespace(r.current)) {
    r.next();
  }
}
function escaped(r) {
  const n = r.next();
  if (n === void 0) {
    throw new Error("Unterminated escape sequence");
  }
  return n;
}
function quoted(r) {
  let s = "";
  const q = r.current;
  const e = q === '"';
  for (let c = r.next(); c !== q; c = r.next()) {
    if (c === void 0) {
      throw new Error(`Unterminated quote: ${q}`);
    }
    if (e && c === "\\") {
      s += escaped(r);
    } else {
      s += c;
    }
  }
  return s;
}
function commandlineParser(line) {
  const args = [];
  let s = "";
  const r = new StringReader(line);
  r.next();
  consumeWhitespace(r);
  let sawQuote = false;
  while (r.current !== void 0) {
    if (isWhitespace(r.current)) {
      args.push(s);
      sawQuote = false;
      s = "";
      consumeWhitespace(r);
    } else if (r.current == "\\") {
      s += escaped(r);
      r.next();
    } else if (r.current === "'" || r.current === '"') {
      s += quoted(r);
      sawQuote = true;
      r.next();
    } else {
      s += r.current;
      r.next();
    }
  }
  if (sawQuote || s.length) {
    args.push(s);
  }
  return args;
}
