// @project-chip/matter.js/dist/cjs/model/elements/DeviceTypeElement.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DeviceTypeElement_exports = {};
__export(DeviceTypeElement_exports, {
  DeviceTypeElement: () => DeviceTypeElement
});
module.exports = __toCommonJS(DeviceTypeElement_exports);
var import_definitions = require("../definitions/index.js");
var import_BaseElement = require("./BaseElement.js");
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
function DeviceTypeElement(definition) {
  return (0, import_BaseElement.BaseElement)(DeviceTypeElement.Tag, definition);
}
((DeviceTypeElement2) => {
  DeviceTypeElement2.Tag = import_definitions.ElementTag.DeviceType;
  let Classification;
  ((Classification2) => {
    Classification2["Base"] = "base";
    Classification2["Node"] = "node";
    Classification2["Utility"] = "utility";
    Classification2["Simple"] = "simple";
    Classification2["Dynamic"] = "dynamic";
  })(Classification = DeviceTypeElement2.Classification || (DeviceTypeElement2.Classification = {}));
})(DeviceTypeElement || (DeviceTypeElement = {}));
