// @project-chip/matter-node.js/dist/esm/time/TimeNode.js
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ImplementationError } from "@project-chip/matter.js/common";
import { Time } from "@project-chip/matter.js/time";
class TimerNode {
  constructor(name, intervalMs, callback, isPeriodic) {
    this.name = name;
    this.intervalMs = intervalMs;
    this.callback = callback;
    this.isPeriodic = isPeriodic;
    if (intervalMs < 0 || intervalMs > 2147483647) {
      throw new ImplementationError(
        `Invalid intervalMs: ${intervalMs}. The value must be between 0 and 32-bit maximum value (2147483647)`
      );
    }
  }
  #timerId;
  #utility = false;
  isRunning = false;
  get systemId() {
    return Number(this.#timerId);
  }
  get utility() {
    return this.#utility;
  }
  set utility(utility) {
    if (utility === this.#utility) {
      return;
    }
    if (utility) {
      this.#timerId?.unref();
    } else {
      this.#timerId?.ref();
    }
    this.#utility = utility;
  }
  start() {
    if (this.isRunning) this.stop();
    Time.register(this);
    this.isRunning = true;
    this.#timerId = (this.isPeriodic ? setInterval : setTimeout)(() => {
      if (!this.isPeriodic) {
        Time.unregister(this);
        this.isRunning = false;
      }
      this.callback();
    }, this.intervalMs);
    return this;
  }
  stop() {
    (this.isPeriodic ? clearInterval : clearTimeout)(this.#timerId);
    Time.unregister(this);
    this.isRunning = false;
    return this;
  }
}
class TimeNode extends Time {
  now() {
    return /* @__PURE__ */ new Date();
  }
  nowMs() {
    return this.now().getTime();
  }
  getTimer(name, durationMs, callback) {
    return new TimerNode(name, durationMs, callback, false);
  }
  getPeriodicTimer(name, intervalMs, callback) {
    return new TimerNode(name, intervalMs, callback, true);
  }
}
export {
  TimeNode
};
