// @project-chip/matter.js/dist/esm/protocol/securechannel/SecureChannelMessages.js
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const SECURE_CHANNEL_PROTOCOL_ID = 0;
var MessageType = /* @__PURE__ */ ((MessageType2) => {
  MessageType2[MessageType2["MsgCounterSyncReq"] = 0] = "MsgCounterSyncReq";
  MessageType2[MessageType2["MsgCounterSyncRsp"] = 1] = "MsgCounterSyncRsp";
  MessageType2[MessageType2["StandaloneAck"] = 16] = "StandaloneAck";
  MessageType2[MessageType2["PbkdfParamRequest"] = 32] = "PbkdfParamRequest";
  MessageType2[MessageType2["PbkdfParamResponse"] = 33] = "PbkdfParamResponse";
  MessageType2[MessageType2["PasePake1"] = 34] = "PasePake1";
  MessageType2[MessageType2["PasePake2"] = 35] = "PasePake2";
  MessageType2[MessageType2["PasePake3"] = 36] = "PasePake3";
  MessageType2[MessageType2["Sigma1"] = 48] = "Sigma1";
  MessageType2[MessageType2["Sigma2"] = 49] = "Sigma2";
  MessageType2[MessageType2["Sigma3"] = 50] = "Sigma3";
  MessageType2[MessageType2["Sigma2Resume"] = 51] = "Sigma2Resume";
  MessageType2[MessageType2["StatusReport"] = 64] = "StatusReport";
  MessageType2[MessageType2["IcdCheckInMessage"] = 80] = "IcdCheckInMessage";
  return MessageType2;
})(MessageType || {});
var ProtocolStatusCode = /* @__PURE__ */ ((ProtocolStatusCode2) => {
  ProtocolStatusCode2[ProtocolStatusCode2["Success"] = 0] = "Success";
  ProtocolStatusCode2[ProtocolStatusCode2["NoSharedTrustRoots"] = 1] = "NoSharedTrustRoots";
  ProtocolStatusCode2[ProtocolStatusCode2["InvalidParam"] = 2] = "InvalidParam";
  ProtocolStatusCode2[ProtocolStatusCode2["CloseSession"] = 3] = "CloseSession";
  ProtocolStatusCode2[ProtocolStatusCode2["Busy"] = 4] = "Busy";
  return ProtocolStatusCode2;
})(ProtocolStatusCode || {});
var GeneralStatusCode = /* @__PURE__ */ ((GeneralStatusCode2) => {
  GeneralStatusCode2[GeneralStatusCode2["Success"] = 0] = "Success";
  GeneralStatusCode2[GeneralStatusCode2["Failure"] = 1] = "Failure";
  GeneralStatusCode2[GeneralStatusCode2["BadPrecondition"] = 2] = "BadPrecondition";
  GeneralStatusCode2[GeneralStatusCode2["OutOfRange"] = 3] = "OutOfRange";
  GeneralStatusCode2[GeneralStatusCode2["BadRequest"] = 4] = "BadRequest";
  GeneralStatusCode2[GeneralStatusCode2["Unsupported"] = 5] = "Unsupported";
  GeneralStatusCode2[GeneralStatusCode2["Unexpected"] = 6] = "Unexpected";
  GeneralStatusCode2[GeneralStatusCode2["ResourceExhausted"] = 7] = "ResourceExhausted";
  GeneralStatusCode2[GeneralStatusCode2["Busy"] = 8] = "Busy";
  GeneralStatusCode2[GeneralStatusCode2["Timeout"] = 9] = "Timeout";
  GeneralStatusCode2[GeneralStatusCode2["Continue"] = 10] = "Continue";
  GeneralStatusCode2[GeneralStatusCode2["Aborted"] = 11] = "Aborted";
  GeneralStatusCode2[GeneralStatusCode2["InvalidArgument"] = 12] = "InvalidArgument";
  GeneralStatusCode2[GeneralStatusCode2["NotFound"] = 13] = "NotFound";
  GeneralStatusCode2[GeneralStatusCode2["AlreadyExists"] = 14] = "AlreadyExists";
  GeneralStatusCode2[GeneralStatusCode2["PermissionDenied"] = 15] = "PermissionDenied";
  GeneralStatusCode2[GeneralStatusCode2["DataLoss"] = 16] = "DataLoss";
  GeneralStatusCode2[GeneralStatusCode2["MessageTooLarge"] = 17] = "MessageTooLarge";
  return GeneralStatusCode2;
})(GeneralStatusCode || {});
export {
  GeneralStatusCode,
  MessageType,
  ProtocolStatusCode,
  SECURE_CHANNEL_PROTOCOL_ID
};
