// @project-chip/matter.js/dist/cjs/cluster/server/OperationalCredentialsServer.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var OperationalCredentialsServer_exports = {};
__export(OperationalCredentialsServer_exports, {
  OperationalCredentialsClusterHandler: () => OperationalCredentialsClusterHandler
});
module.exports = __toCommonJS(OperationalCredentialsServer_exports);
var import_DeviceCertification = require("../../behavior/definitions/operational-credentials/DeviceCertification.js");
var import_OperationalCredentialsTypes = require("../../behavior/definitions/operational-credentials/OperationalCredentialsTypes.js");
var import_CertificateManager = require("../../certificate/CertificateManager.js");
var import_FailsafeContext = require("../../common/FailsafeContext.js");
var import_FailsafeTimer = require("../../common/FailsafeTimer.js");
var import_MatterError = require("../../common/MatterError.js");
var import_ValidationError = require("../../common/ValidationError.js");
var import_Crypto = require("../../crypto/Crypto.js");
var import_FabricIndex = require("../../datatype/FabricIndex.js");
var import_Fabric = require("../../fabric/Fabric.js");
var import_FabricManager = require("../../fabric/FabricManager.js");
var import_Logger = require("../../log/Logger.js");
var import_StatusCode = require("../../protocol/interaction/StatusCode.js");
var import_SecureSession = require("../../session/SecureSession.js");
var import_TlvBoolean = require("../../tlv/TlvBoolean.js");
var import_TlvObject = require("../../tlv/TlvObject.js");
var import_TlvString = require("../../tlv/TlvString.js");
var import_Cluster = require("../Cluster.js");
var import_BasicInformationCluster = require("../definitions/BasicInformationCluster.js");
var import_OperationalCredentialsCluster = require("../definitions/OperationalCredentialsCluster.js");
var import_definitions = require("../definitions/index.js");
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const logger = import_Logger.Logger.get("OperationalCredentialsServer");
import_OperationalCredentialsCluster.OperationalCredentials.Cluster.commands = {
  ...import_OperationalCredentialsCluster.OperationalCredentials.Cluster.commands,
  attestationRequest: (0, import_Cluster.Command)(
    0,
    (0, import_TlvObject.TlvObject)({ attestationNonce: (0, import_TlvObject.TlvField)(0, import_TlvString.TlvByteString) }),
    1,
    import_OperationalCredentialsCluster.OperationalCredentials.TlvAttestationResponse,
    { invokeAcl: import_Cluster.AccessLevel.Administer }
  ),
  csrRequest: (0, import_Cluster.Command)(
    4,
    (0, import_TlvObject.TlvObject)({
      csrNonce: (0, import_TlvObject.TlvField)(0, import_TlvString.TlvByteString),
      isForUpdateNoc: (0, import_TlvObject.TlvOptionalField)(1, import_TlvBoolean.TlvBoolean)
    }),
    5,
    import_OperationalCredentialsCluster.OperationalCredentials.TlvCsrResponse,
    { invokeAcl: import_Cluster.AccessLevel.Administer }
  )
};
const OperationalCredentialsClusterHandler = (certificates, productDescription) => {
  let certification = void 0;
  const cert = () => {
    if (certification !== void 0) {
      return certification;
    }
    return certification = new import_DeviceCertification.DeviceCertification(certificates, productDescription);
  };
  const assureCertification = async () => {
    await cert().construction;
  };
  return {
    attestationRequest: async ({ request: { attestationNonce }, session }) => {
      if (attestationNonce.length !== 32) {
        throw new import_StatusCode.StatusResponseError("Invalid attestation nonce length", import_StatusCode.StatusCode.InvalidCommand);
      }
      (0, import_SecureSession.assertSecureSession)(session);
      await assureCertification();
      const elements = import_OperationalCredentialsTypes.TlvAttestation.encode({
        declaration: cert().declaration,
        attestationNonce,
        timestamp: 0
      });
      return { attestationElements: elements, attestationSignature: cert().sign(session, elements) };
    },
    csrRequest: async ({ request: { csrNonce, isForUpdateNoc }, session }) => {
      if (csrNonce.length !== 32) {
        throw new import_StatusCode.StatusResponseError("Invalid CSR nonce length", import_StatusCode.StatusCode.InvalidCommand);
      }
      (0, import_SecureSession.assertSecureSession)(session);
      if (isForUpdateNoc && session.isPase) {
        throw new import_StatusCode.StatusResponseError(
          "csrRequest for UpdateNoc received on a PASE session.",
          import_StatusCode.StatusCode.InvalidCommand
        );
      }
      const device = session.context;
      device.assertFailSafeArmed("csrRequest received while failsafe is not armed.");
      const timedOp = device.failsafeContext;
      if (timedOp.fabricIndex !== void 0) {
        throw new import_StatusCode.StatusResponseError(
          `csrRequest received after ${timedOp.forUpdateNoc ? "UpdateNOC" : "AddNOC"} already invoked.`,
          import_StatusCode.StatusCode.ConstraintError
        );
      }
      await assureCertification();
      const certSigningRequest = timedOp.createCertificateSigningRequest(isForUpdateNoc ?? false, session.id);
      const nocsrElements = import_OperationalCredentialsTypes.TlvCertSigningRequest.encode({ certSigningRequest, csrNonce });
      return { nocsrElements, attestationSignature: cert().sign(session, nocsrElements) };
    },
    certificateChainRequest: async ({ request: { certificateType } }) => {
      await assureCertification();
      switch (certificateType) {
        case import_OperationalCredentialsCluster.OperationalCredentials.CertificateChainType.DacCertificate:
          return { certificate: cert().certificate };
        case import_OperationalCredentialsCluster.OperationalCredentials.CertificateChainType.PaiCertificate:
          return { certificate: cert().intermediateCertificate };
        default:
          throw new import_StatusCode.StatusResponseError(
            `Unsupported certificate type: ${certificateType}`,
            import_StatusCode.StatusCode.InvalidCommand
          );
      }
    },
    addNoc: async ({
      request: { nocValue, icacValue, ipkValue, caseAdminSubject, adminVendorId },
      attributes: { nocs, commissionedFabrics, fabrics, trustedRootCertificates, supportedFabrics },
      session,
      endpoint,
      message
    }) => {
      if (!session.isSecure)
        throw new import_MatterError.MatterFlowError("addOperationalCert should be called on a secure session.");
      const device = session.context;
      const failsafeContext = device.failsafeContext;
      if (failsafeContext.fabricIndex !== void 0) {
        throw new import_StatusCode.StatusResponseError(
          `addNoc received after ${failsafeContext.forUpdateNoc ? "UpdateNOC" : "AddNOC"} already invoked.`,
          import_StatusCode.StatusCode.ConstraintError
        );
      }
      if (!failsafeContext.hasRootCert) {
        return {
          statusCode: import_OperationalCredentialsCluster.OperationalCredentials.NodeOperationalCertStatus.InvalidNoc,
          debugText: "Root certificate not found."
        };
      }
      if (failsafeContext.csrSessionId !== session.id) {
        return {
          statusCode: import_OperationalCredentialsCluster.OperationalCredentials.NodeOperationalCertStatus.MissingCsr,
          debugText: "CSR not found in failsafe context."
        };
      }
      if (failsafeContext.forUpdateNoc) {
        throw new import_StatusCode.StatusResponseError(
          `addNoc received after csr request was invoked for UpdateNOC.`,
          import_StatusCode.StatusCode.ConstraintError
        );
      }
      if (device.getFabrics().length === supportedFabrics.getLocal()) {
        return {
          statusCode: import_OperationalCredentialsCluster.OperationalCredentials.NodeOperationalCertStatus.TableFull,
          debugText: `No more fabrics can be added because limit ${supportedFabrics.getLocal()} reached.`
        };
      }
      let fabric;
      try {
        fabric = await failsafeContext.buildFabric({
          nocValue,
          icacValue,
          adminVendorId,
          ipkValue,
          caseAdminSubject
        });
      } catch (error) {
        logger.info("building fabric failed", error);
        if (error instanceof import_FailsafeTimer.MatterFabricConflictError) {
          return {
            statusCode: import_OperationalCredentialsCluster.OperationalCredentials.NodeOperationalCertStatus.FabricConflict,
            debugText: error.message
          };
        } else if (error instanceof import_FabricManager.FabricTableFullError) {
          return {
            statusCode: import_OperationalCredentialsCluster.OperationalCredentials.NodeOperationalCertStatus.TableFull,
            debugText: error.message
          };
        } else if (error instanceof import_Crypto.CryptoVerifyError || error instanceof import_CertificateManager.CertificateError || error instanceof import_ValidationError.ValidationError || error instanceof import_MatterError.UnexpectedDataError) {
          return {
            statusCode: import_OperationalCredentialsCluster.OperationalCredentials.NodeOperationalCertStatus.InvalidNoc,
            debugText: error.message
          };
        } else if (error instanceof import_Fabric.PublicKeyError) {
          return {
            statusCode: import_OperationalCredentialsCluster.OperationalCredentials.NodeOperationalCertStatus.InvalidPublicKey,
            debugText: error.message
          };
        } else if (error instanceof import_FailsafeContext.MatterFabricInvalidAdminSubjectError) {
          return {
            statusCode: import_OperationalCredentialsCluster.OperationalCredentials.NodeOperationalCertStatus.InvalidAdminSubject,
            debugText: error.message
          };
        }
        throw error;
      }
      await failsafeContext.addFabric(fabric);
      (0, import_SecureSession.assertSecureSession)(session);
      if (session.isPase) {
        logger.debug(`Add Fabric with index ${fabric.fabricIndex} to PASE session ${session.name}.`);
        session.addAssociatedFabric(fabric);
      }
      nocs.updated(session);
      commissionedFabrics.updated(session);
      fabrics.updated(session);
      trustedRootCertificates.updated(session);
      const aclServer = endpoint.getClusterServer(import_definitions.AccessControl.Cluster);
      if (aclServer !== void 0) {
        aclServer.attributes.acl.set(
          [
            {
              fabricIndex: fabric.fabricIndex,
              privilege: import_definitions.AccessControl.AccessControlEntryPrivilege.Administer,
              authMode: import_definitions.AccessControl.AccessControlEntryAuthMode.Case,
              subjects: [caseAdminSubject],
              targets: null
              // entire node
            }
          ],
          session,
          message,
          false,
          true
        );
      }
      logger.info(`addNoc success, adminVendorId ${adminVendorId}, caseAdminSubject ${caseAdminSubject}`);
      return { statusCode: import_OperationalCredentialsCluster.OperationalCredentials.NodeOperationalCertStatus.Ok, fabricIndex: fabric.fabricIndex };
    },
    fabricsAttributeGetter: ({ session, isFabricFiltered }) => {
      if (session === void 0 || !session.isSecure) return [];
      const fabrics = isFabricFiltered ? [session.associatedFabric] : session.context.getFabrics();
      return fabrics.map((fabric) => ({
        fabricId: fabric.fabricId,
        label: fabric.label,
        nodeId: fabric.nodeId,
        rootPublicKey: fabric.rootPublicKey,
        vendorId: fabric.rootVendorId,
        fabricIndex: fabric.fabricIndex
      }));
    },
    // Needed because FabricScopedAttributeServer clas requires both getter and setter if custom
    fabricsAttributeSetter: () => {
      throw new import_MatterError.MatterFlowError("fabrics attribute is read-only.");
    },
    nocsAttributeGetter: ({ session, isFabricFiltered }) => {
      if (session === void 0 || !session.isSecure) return [];
      const fabrics = isFabricFiltered ? [session.associatedFabric] : session.context.getFabrics();
      return fabrics.map((fabric) => ({
        noc: fabric.operationalCert,
        icac: fabric.intermediateCACert ?? null,
        fabricIndex: fabric.fabricIndex
      }));
    },
    // Needed because FabricScopedAttributeServer clas requires both getter and setter if custom
    nocsAttributeSetter: () => {
      throw new import_MatterError.MatterFlowError("fabrics attribute is read-only.");
    },
    commissionedFabricsAttributeGetter: ({ session }) => {
      if (session === void 0 || !session.isSecure) return 0;
      return session.context.getFabrics().length;
    },
    trustedRootCertificatesAttributeGetter: ({ session }) => {
      if (session === void 0 || !session.isSecure) {
        logger.debug(`trustedRootCertificatesAttributeGetter: session not set or not secure ${!!session}`);
        return [];
      }
      if (!session.isSecure)
        throw new import_MatterError.MatterFlowError("addOperationalCert should be called on a secure session.");
      const rootCerts = session.context.getFabrics().map((fabric) => fabric.rootCert);
      const device = session.context;
      if (device.isFailsafeArmed()) {
        const failsafeContext = device.failsafeContext;
        const temporaryRootCert = failsafeContext.rootCert;
        if (temporaryRootCert !== void 0) {
          logger.debug(`Add temporary trusted root certificate to the list.`);
          rootCerts.push(temporaryRootCert);
        } else {
          logger.debug(`No temporary trusted root certificate to be added.`);
        }
      }
      return rootCerts;
    },
    currentFabricIndexAttributeGetter: ({ session }) => {
      if (session === void 0 || !session.isSecure) return import_FabricIndex.FabricIndex.NO_FABRIC;
      (0, import_SecureSession.assertSecureSession)(session);
      return session.fabric?.fabricIndex ?? import_FabricIndex.FabricIndex.NO_FABRIC;
    },
    updateNoc: async ({ request: { nocValue, icacValue }, attributes: { nocs, fabrics }, session }) => {
      (0, import_SecureSession.assertSecureSession)(session);
      session.associatedFabric;
      const device = session.context;
      const failsafeContext = device.failsafeContext;
      if (failsafeContext.fabricIndex !== void 0) {
        throw new import_StatusCode.StatusResponseError(
          `updateNoc received after ${failsafeContext.forUpdateNoc ? "UpdateNOC" : "AddNOC"} already invoked.`,
          import_StatusCode.StatusCode.ConstraintError
        );
      }
      if (failsafeContext.forUpdateNoc) {
        throw new import_StatusCode.StatusResponseError(
          `addNoc received after csr request was invoked for UpdateNOC.`,
          import_StatusCode.StatusCode.ConstraintError
        );
      }
      if (failsafeContext.hasRootCert) {
        throw new import_StatusCode.StatusResponseError(
          "Trusted root certificate added in this session which is now allowed for UpdateNOC.",
          import_StatusCode.StatusCode.ConstraintError
        );
      }
      if (!failsafeContext.forUpdateNoc) {
        throw new import_StatusCode.StatusResponseError("csrRequest not invoked for UpdateNOC.", import_StatusCode.StatusCode.ConstraintError);
      }
      if (session.associatedFabric.fabricIndex !== failsafeContext.associatedFabric?.fabricIndex) {
        throw new import_StatusCode.StatusResponseError(
          "Fabric of this session and the failsafe context do not match.",
          import_StatusCode.StatusCode.ConstraintError
        );
      }
      try {
        const updateFabric = await failsafeContext.buildUpdatedFabric(nocValue, icacValue);
        await failsafeContext.updateFabric(updateFabric);
        nocs.updated(session);
        fabrics.updated(session);
        return {
          statusCode: import_OperationalCredentialsCluster.OperationalCredentials.NodeOperationalCertStatus.Ok,
          fabricIndex: updateFabric.fabricIndex
        };
      } catch (error) {
        logger.info("building fabric for update failed", error);
        if (error instanceof import_Crypto.CryptoVerifyError || error instanceof import_CertificateManager.CertificateError || error instanceof import_ValidationError.ValidationError || error instanceof import_MatterError.UnexpectedDataError) {
          return {
            statusCode: import_OperationalCredentialsCluster.OperationalCredentials.NodeOperationalCertStatus.InvalidNoc,
            debugText: error.message
          };
        } else if (error instanceof import_Fabric.PublicKeyError) {
          return {
            statusCode: import_OperationalCredentialsCluster.OperationalCredentials.NodeOperationalCertStatus.InvalidPublicKey,
            debugText: error.message
          };
        }
        throw error;
      }
    },
    updateFabricLabel: async ({ request: { label }, attributes: { fabrics }, session }) => {
      (0, import_SecureSession.assertSecureSession)(session, "updateOperationalCert should be called on a secure session.");
      const fabric = session.associatedFabric;
      const currentFabricIndex = fabric.fabricIndex;
      const device = session.context;
      const conflictingLabelFabric = device.getFabrics().find((f) => f.label === label && f.fabricIndex !== currentFabricIndex);
      if (conflictingLabelFabric !== void 0) {
        return {
          statusCode: import_OperationalCredentialsCluster.OperationalCredentials.NodeOperationalCertStatus.LabelConflict,
          debugText: `Label ${label} already used by fabric ${conflictingLabelFabric.fabricIndex}`
        };
      }
      await fabric.setLabel(label);
      fabrics.updated(session);
      return { statusCode: import_OperationalCredentialsCluster.OperationalCredentials.NodeOperationalCertStatus.Ok, fabricIndex: fabric.fabricIndex };
    },
    removeFabric: async ({
      request: { fabricIndex },
      attributes: { nocs, commissionedFabrics, fabrics, trustedRootCertificates },
      session,
      endpoint
    }) => {
      const device = session.context;
      const fabric = device.getFabricByIndex(fabricIndex);
      if (fabric === void 0) {
        return {
          statusCode: import_OperationalCredentialsCluster.OperationalCredentials.NodeOperationalCertStatus.InvalidFabricIndex,
          debugText: `Fabric ${fabricIndex} not found`
        };
      }
      const basicInformationCluster = endpoint.getClusterServer(import_BasicInformationCluster.BasicInformation.Cluster);
      basicInformationCluster?.triggerLeaveEvent?.({ fabricIndex });
      (0, import_SecureSession.assertSecureSession)(session);
      await fabric.remove(session.id);
      nocs.updated(session);
      commissionedFabrics.updated(session);
      fabrics.updated(session);
      trustedRootCertificates.updated(session);
      return {
        statusCode: import_OperationalCredentialsCluster.OperationalCredentials.NodeOperationalCertStatus.Ok,
        fabricIndex
      };
    },
    addTrustedRootCertificate: async ({
      request: { rootCaCertificate },
      attributes: { trustedRootCertificates },
      session
    }) => {
      (0, import_SecureSession.assertSecureSession)(session);
      const failsafeContext = session.context.failsafeContext;
      if (failsafeContext.hasRootCert) {
        throw new import_StatusCode.StatusResponseError(
          "Trusted root certificate already added in this FailSafe context.",
          import_StatusCode.StatusCode.ConstraintError
        );
      }
      if (failsafeContext.fabricIndex !== void 0) {
        throw new import_StatusCode.StatusResponseError(
          `Cannot add trusted root certificates after ${failsafeContext.forUpdateNoc ? "UpdateNOC" : "AddNOC"}.`,
          import_StatusCode.StatusCode.ConstraintError
        );
      }
      try {
        failsafeContext.setRootCert(rootCaCertificate);
      } catch (error) {
        logger.info("setting root certificate failed", error);
        if (error instanceof import_Crypto.CryptoVerifyError || error instanceof import_CertificateManager.CertificateError || error instanceof import_ValidationError.ValidationError || error instanceof import_MatterError.UnexpectedDataError) {
          throw new import_StatusCode.StatusResponseError(error.message, import_StatusCode.StatusCode.InvalidCommand);
        }
        throw error;
      }
      trustedRootCertificates.updated(session);
    }
  };
};
