// @project-chip/matter.js/dist/esm/model/standard/elements/AirPurifierDT.js
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Matter } from "../Matter.js";
import { DeviceTypeElement as DeviceType, RequirementElement as Requirement } from "../../elements/index.js";
const AirPurifierDt = DeviceType({
  name: "AirPurifier",
  id: 45,
  category: "HVAC",
  classification: "simple",
  details: "An Air Purifier is a standalone device that is designed to clean the air in a room.\nIt is a device that has a fan to control the air speed while it is operating. Optionally, it can report on the condition of its filters.",
  xref: { document: "device", section: "9.3" },
  children: [
    Requirement({
      name: "Descriptor",
      id: 29,
      element: "serverCluster",
      children: [
        Requirement({ name: "DeviceTypeList", default: [{ deviceType: 45, revision: 1 }], element: "attribute" })
      ]
    }),
    Requirement({
      name: "Identify",
      id: 3,
      conformance: "M",
      element: "serverCluster",
      xref: { document: "device", section: "9.3.5" }
    }),
    Requirement({
      name: "Groups",
      id: 4,
      conformance: "O",
      element: "serverCluster",
      xref: { document: "device", section: "9.3.5" }
    }),
    Requirement({
      name: "FanControl",
      id: 514,
      conformance: "M",
      element: "serverCluster",
      xref: { document: "device", section: "9.3.5" }
    }),
    Requirement({
      name: "HepaFilterMonitoring",
      id: 113,
      conformance: "O",
      element: "serverCluster",
      xref: { document: "device", section: "9.3.5" }
    }),
    Requirement({
      name: "ActivatedCarbonFilterMonitoring",
      id: 114,
      conformance: "O",
      element: "serverCluster",
      xref: { document: "device", section: "9.3.5" }
    })
  ]
});
Matter.children.push(AirPurifierDt);
export {
  AirPurifierDt
};
