// @project-chip/matter.js/dist/esm/model/standard/elements/CompassDirectionNS.js
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Matter } from "../Matter.js";
import {
  SemanticNamespaceElement as SemanticNamespace,
  SemanticTagElement as SemanticTag
} from "../../elements/index.js";
const CompassDirectionNs = SemanticNamespace({
  name: "CompassDirection",
  id: 2,
  mfgCode: void 0,
  details: "The tags contained in this namespace may be used in any domain or context, to indicate an association with a movement into a certain compass direction. Note the difference with Chapter 4, Common Compass Location Semantic Tag Namespace.",
  xref: { document: "namespace", section: "3" },
  children: [
    SemanticTag({ name: "Northward", id: 0 }),
    SemanticTag({ name: "North-Eastward", id: 1 }),
    SemanticTag({ name: "Eastward", id: 2 }),
    SemanticTag({ name: "South-Eastward", id: 3 }),
    SemanticTag({ name: "Southward", id: 4 }),
    SemanticTag({ name: "South-Westward", id: 5 }),
    SemanticTag({ name: "Westward", id: 6 }),
    SemanticTag({ name: "North-Westward", id: 7 })
  ]
});
Matter.children.push(CompassDirectionNs);
export {
  CompassDirectionNs
};
