// @project-chip/matter.js/dist/esm/model/standard/elements/ContactSensorDT.js
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Matter } from "../Matter.js";
import { DeviceTypeElement as DeviceType, RequirementElement as Requirement } from "../../elements/index.js";
const ContactSensorDt = DeviceType({
  name: "ContactSensor",
  id: 21,
  category: "Sensor",
  classification: "simple",
  details: "This defines conformance to the Contact Sensor device type.",
  xref: { document: "device", section: "7.1" },
  children: [
    Requirement({
      name: "Descriptor",
      id: 29,
      element: "serverCluster",
      children: [
        Requirement({ name: "DeviceTypeList", default: [{ deviceType: 21, revision: 2 }], element: "attribute" })
      ]
    }),
    Requirement({
      name: "Identify",
      id: 3,
      conformance: "M",
      element: "serverCluster",
      xref: { document: "device", section: "7.1.4" }
    }),
    Requirement({
      name: "BooleanState",
      id: 69,
      conformance: "M",
      element: "serverCluster",
      xref: { document: "device", section: "7.1.4" }
    }),
    Requirement({
      name: "BooleanStateConfiguration",
      id: 128,
      conformance: "O",
      element: "serverCluster",
      xref: { document: "device", section: "7.1.4" }
    })
  ]
});
Matter.children.push(ContactSensorDt);
export {
  ContactSensorDt
};
