// @project-chip/matter.js/dist/esm/model/standard/elements/CooktopDT.js
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Matter } from "../Matter.js";
import { DeviceTypeElement as DeviceType, RequirementElement as Requirement } from "../../elements/index.js";
const CooktopDt = DeviceType({
  name: "Cooktop",
  id: 120,
  category: "Appliances",
  classification: "simple",
  details: "A cooktop is a cooking surface that heats food either by transferring currents from an electromagnetic field located below the glass surface directly to the magnetic induction cookware placed above or through traditional gas or electric burners.",
  xref: { document: "device", section: "13.8" },
  children: [
    Requirement({
      name: "Descriptor",
      id: 29,
      element: "serverCluster",
      children: [Requirement({ name: "DeviceTypeList", default: [{ deviceType: 120, revision: 1 }], element: "attribute" })]
    }),
    Requirement({
      name: "Identify",
      id: 3,
      conformance: "O",
      element: "serverCluster",
      xref: { document: "device", section: "13.8.5" }
    }),
    Requirement({
      name: "OnOff",
      id: 6,
      conformance: "M",
      element: "serverCluster",
      xref: { document: "device", section: "13.8.5" },
      children: [Requirement({ name: "OFFONLY", conformance: "M", element: "feature" })]
    })
  ]
});
Matter.children.push(CooktopDt);
export {
  CooktopDt
};
