// @project-chip/matter.js/dist/esm/model/standard/elements/FanDT.js
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Matter } from "../Matter.js";
import { DeviceTypeElement as DeviceType, RequirementElement as Requirement } from "../../elements/index.js";
const FanDt = DeviceType({
  name: "Fan",
  id: 43,
  category: "HVAC",
  classification: "simple",
  details: "A Fan device is typically standalone or mounted on a ceiling or wall and is used to circulate air in a room.",
  xref: { document: "device", section: "9.2" },
  children: [
    Requirement({
      name: "Descriptor",
      id: 29,
      element: "serverCluster",
      children: [
        Requirement({ name: "DeviceTypeList", default: [{ deviceType: 43, revision: 2 }], element: "attribute" })
      ]
    }),
    Requirement({
      name: "Identify",
      id: 3,
      conformance: "M",
      element: "serverCluster",
      xref: { document: "device", section: "9.2.5" }
    }),
    Requirement({
      name: "Groups",
      id: 4,
      conformance: "M",
      element: "serverCluster",
      xref: { document: "device", section: "9.2.5" }
    }),
    Requirement({
      name: "FanControl",
      id: 514,
      conformance: "M",
      element: "serverCluster",
      xref: { document: "device", section: "9.2.5" },
      children: [
        Requirement({ name: "FanModeSequence", access: "R V", conformance: "Matter", element: "attribute" })
      ]
    })
  ]
});
Matter.children.push(FanDt);
export {
  FanDt
};
