// @project-chip/matter.js/dist/esm/model/standard/elements/GenericSwitchDT.js
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Matter } from "../Matter.js";
import { DeviceTypeElement as DeviceType, RequirementElement as Requirement } from "../../elements/index.js";
const GenericSwitchDt = DeviceType({
  name: "GenericSwitch",
  id: 15,
  category: "Switches and Controls",
  classification: "simple",
  details: "This defines conformance for the Generic Switch device type.",
  xref: { document: "device", section: "6.6" },
  children: [
    Requirement({
      name: "Descriptor",
      id: 29,
      element: "serverCluster",
      children: [
        Requirement({ name: "DeviceTypeList", default: [{ deviceType: 15, revision: 3 }], element: "attribute" })
      ]
    }),
    Requirement({
      name: "Identify",
      id: 3,
      conformance: "M",
      element: "serverCluster",
      xref: { document: "device", section: "6.6.4" }
    }),
    Requirement({
      name: "Switch",
      id: 59,
      conformance: "M",
      element: "serverCluster",
      xref: { document: "device", section: "6.6.4" }
    })
  ]
});
Matter.children.push(GenericSwitchDt);
export {
  GenericSwitchDt
};
