// @project-chip/matter.js/dist/esm/model/standard/elements/LightSensorDT.js
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Matter } from "../Matter.js";
import { DeviceTypeElement as DeviceType, RequirementElement as Requirement } from "../../elements/index.js";
const LightSensorDt = DeviceType({
  name: "LightSensor",
  id: 262,
  category: "Sensor",
  classification: "simple",
  details: "A Light Sensor device is a measurement and sensing device that is capable of measuring and reporting the intensity of light (illuminance) to which the sensor is being subjected.",
  xref: { document: "device", section: "7.2" },
  children: [
    Requirement({
      name: "Descriptor",
      id: 29,
      element: "serverCluster",
      children: [Requirement({ name: "DeviceTypeList", default: [{ deviceType: 262, revision: 3 }], element: "attribute" })]
    }),
    Requirement({
      name: "Identify",
      id: 3,
      conformance: "M",
      element: "serverCluster",
      xref: { document: "device", section: "7.2.4" }
    }),
    Requirement({
      name: "IlluminanceMeasurement",
      id: 1024,
      conformance: "M",
      element: "serverCluster",
      xref: { document: "device", section: "7.2.4" }
    })
  ]
});
Matter.children.push(LightSensorDt);
export {
  LightSensorDt
};
