// @project-chip/matter.js/dist/esm/model/standard/elements/RefrigeratorAlarm.js
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Matter } from "../Matter.js";
import {
  ClusterElement as Cluster,
  AttributeElement as Attribute,
  FieldElement as Field,
  CommandElement as Command,
  DatatypeElement as Datatype
} from "../../elements/index.js";
const RefrigeratorAlarm = Cluster({
  name: "RefrigeratorAlarm",
  id: 87,
  type: "AlarmBase",
  classification: "application",
  pics: "REFALM",
  details: "This cluster is a derived cluster of Alarm Base cluster.",
  xref: { document: "cluster", section: "8.8" },
  children: [
    Attribute({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
    Attribute({
      name: "FeatureMap",
      id: 65532,
      type: "FeatureMap",
      xref: { document: "cluster", section: "8.8.4" },
      children: [Field({
        name: "RESET",
        conformance: "X",
        constraint: "0",
        description: "Reset",
        details: "Supports the ability to reset alarms"
      })]
    }),
    Command({
      name: "ModifyEnabledAlarms",
      id: 1,
      conformance: "X",
      xref: { document: "cluster", section: "8.8.7" }
    }),
    Datatype({
      name: "AlarmBitmap",
      type: "map32",
      xref: { document: "cluster", section: "8.8.5.1" },
      children: [Field({
        name: "DoorOpen",
        constraint: "0",
        description: "The cabinet\u2019s door has been open for a vendor defined amount of time."
      })]
    })
  ]
});
Matter.children.push(RefrigeratorAlarm);
export {
  RefrigeratorAlarm
};
