// @project-chip/matter.js/dist/esm/model/standard/elements/RoomAirConditionerDT.js
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Matter } from "../Matter.js";
import { DeviceTypeElement as DeviceType, RequirementElement as Requirement } from "../../elements/index.js";
const RoomAirConditionerDt = DeviceType({
  name: "RoomAirConditioner",
  id: 114,
  category: "Appliances",
  classification: "simple",
  details: "This defines conformance to the Room Air Conditioner device type.\nA Room Air Conditioner is a device with the primary function of controlling the air temperature in a single room.",
  xref: { document: "device", section: "13.3" },
  children: [
    Requirement({
      name: "Descriptor",
      id: 29,
      element: "serverCluster",
      children: [Requirement({ name: "DeviceTypeList", default: [{ deviceType: 114, revision: 2 }], element: "attribute" })]
    }),
    Requirement({
      name: "Identify",
      id: 3,
      conformance: "M",
      element: "serverCluster",
      xref: { document: "device", section: "13.3.6" }
    }),
    Requirement({
      name: "Groups",
      id: 4,
      conformance: "O",
      element: "serverCluster",
      xref: { document: "device", section: "13.3.6" }
    }),
    Requirement({
      name: "ScenesManagement",
      id: 98,
      conformance: "P, O",
      element: "serverCluster",
      xref: { document: "device", section: "13.3.6" }
    }),
    Requirement({
      name: "OnOff",
      id: 6,
      conformance: "M",
      element: "serverCluster",
      xref: { document: "device", section: "13.3.6" },
      children: [Requirement({ name: "DEADFRONTBEHAVIOR", conformance: "M", element: "feature" })]
    }),
    Requirement({
      name: "Thermostat",
      id: 513,
      conformance: "M",
      element: "serverCluster",
      xref: { document: "device", section: "13.3.6" }
    }),
    Requirement({
      name: "FanControl",
      id: 514,
      conformance: "O",
      element: "serverCluster",
      xref: { document: "device", section: "13.3.6" }
    }),
    Requirement({
      name: "ThermostatUserInterfaceConfiguration",
      id: 516,
      conformance: "O",
      element: "serverCluster",
      xref: { document: "device", section: "13.3.6" },
      children: [Requirement({ name: "KeypadLockout", conformance: "O", element: "attribute" })]
    }),
    Requirement({
      name: "TemperatureMeasurement",
      id: 1026,
      conformance: "O",
      element: "serverCluster",
      xref: { document: "device", section: "13.3.6" }
    }),
    Requirement({
      name: "RelativeHumidityMeasurement",
      id: 1029,
      conformance: "O",
      element: "serverCluster",
      xref: { document: "device", section: "13.3.6" }
    })
  ]
});
Matter.children.push(RoomAirConditionerDt);
export {
  RoomAirConditionerDt
};
