// @project-chip/matter.js/dist/esm/endpoint/definitions/device/ColorDimmerSwitchDevice.js
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { IdentifyServer as BaseIdentifyServer } from "../../../behavior/definitions/identify/IdentifyServer.js";
import { IdentifyBehavior as BaseIdentifyBehavior } from "../../../behavior/definitions/identify/IdentifyBehavior.js";
import { OnOffBehavior as BaseOnOffBehavior } from "../../../behavior/definitions/on-off/OnOffBehavior.js";
import {
  LevelControlBehavior as BaseLevelControlBehavior
} from "../../../behavior/definitions/level-control/LevelControlBehavior.js";
import {
  ColorControlBehavior as BaseColorControlBehavior
} from "../../../behavior/definitions/color-control/ColorControlBehavior.js";
import { GroupsBehavior as BaseGroupsBehavior } from "../../../behavior/definitions/groups/GroupsBehavior.js";
import {
  ScenesManagementBehavior as BaseScenesManagementBehavior
} from "../../../behavior/definitions/scenes-management/ScenesManagementBehavior.js";
import { MutableEndpoint } from "../../type/MutableEndpoint.js";
import { SupportedBehaviors } from "../../properties/SupportedBehaviors.js";
var ColorDimmerSwitchRequirements;
((ColorDimmerSwitchRequirements2) => {
  ColorDimmerSwitchRequirements2.IdentifyServer = BaseIdentifyServer;
  ColorDimmerSwitchRequirements2.IdentifyBehavior = BaseIdentifyBehavior;
  ColorDimmerSwitchRequirements2.OnOffBehavior = BaseOnOffBehavior;
  ColorDimmerSwitchRequirements2.LevelControlBehavior = BaseLevelControlBehavior;
  ColorDimmerSwitchRequirements2.ColorControlBehavior = BaseColorControlBehavior;
  ColorDimmerSwitchRequirements2.GroupsBehavior = BaseGroupsBehavior;
  ColorDimmerSwitchRequirements2.ScenesManagementBehavior = BaseScenesManagementBehavior;
  ColorDimmerSwitchRequirements2.server = { mandatory: { Identify: ColorDimmerSwitchRequirements2.IdentifyServer } };
  ColorDimmerSwitchRequirements2.client = {
    mandatory: {
      Identify: ColorDimmerSwitchRequirements2.IdentifyBehavior,
      OnOff: ColorDimmerSwitchRequirements2.OnOffBehavior,
      LevelControl: ColorDimmerSwitchRequirements2.LevelControlBehavior,
      ColorControl: ColorDimmerSwitchRequirements2.ColorControlBehavior
    },
    optional: { Groups: ColorDimmerSwitchRequirements2.GroupsBehavior, ScenesManagement: ColorDimmerSwitchRequirements2.ScenesManagementBehavior }
  };
})(ColorDimmerSwitchRequirements || (ColorDimmerSwitchRequirements = {}));
const ColorDimmerSwitchDeviceDefinition = MutableEndpoint({
  name: "ColorDimmerSwitch",
  deviceType: 261,
  deviceRevision: 3,
  requirements: ColorDimmerSwitchRequirements,
  behaviors: SupportedBehaviors(ColorDimmerSwitchRequirements.server.mandatory.Identify)
});
const ColorDimmerSwitchDevice = ColorDimmerSwitchDeviceDefinition;
export {
  ColorDimmerSwitchDevice,
  ColorDimmerSwitchDeviceDefinition,
  ColorDimmerSwitchRequirements
};
