// @project-chip/matter.js/dist/esm/endpoint/definitions/device/ContentAppDevice.js
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  KeypadInputServer as BaseKeypadInputServer
} from "../../../behavior/definitions/keypad-input/KeypadInputServer.js";
import {
  ApplicationLauncherServer as BaseApplicationLauncherServer
} from "../../../behavior/definitions/application-launcher/ApplicationLauncherServer.js";
import {
  ApplicationBasicServer as BaseApplicationBasicServer
} from "../../../behavior/definitions/application-basic/ApplicationBasicServer.js";
import { BindingServer as BaseBindingServer } from "../../../behavior/definitions/binding/BindingServer.js";
import { ChannelServer as BaseChannelServer } from "../../../behavior/definitions/channel/ChannelServer.js";
import {
  TargetNavigatorServer as BaseTargetNavigatorServer
} from "../../../behavior/definitions/target-navigator/TargetNavigatorServer.js";
import {
  MediaPlaybackServer as BaseMediaPlaybackServer
} from "../../../behavior/definitions/media-playback/MediaPlaybackServer.js";
import {
  ContentLauncherServer as BaseContentLauncherServer
} from "../../../behavior/definitions/content-launcher/ContentLauncherServer.js";
import {
  AccountLoginServer as BaseAccountLoginServer
} from "../../../behavior/definitions/account-login/AccountLoginServer.js";
import {
  ContentAppObserverBehavior as BaseContentAppObserverBehavior
} from "../../../behavior/definitions/content-app-observer/ContentAppObserverBehavior.js";
import { MutableEndpoint } from "../../type/MutableEndpoint.js";
import { SupportedBehaviors } from "../../properties/SupportedBehaviors.js";
var ContentAppRequirements;
((ContentAppRequirements2) => {
  ContentAppRequirements2.KeypadInputServer = BaseKeypadInputServer;
  ContentAppRequirements2.ApplicationLauncherServer = BaseApplicationLauncherServer;
  ContentAppRequirements2.ApplicationBasicServer = BaseApplicationBasicServer;
  ContentAppRequirements2.BindingServer = BaseBindingServer;
  ContentAppRequirements2.ChannelServer = BaseChannelServer;
  ContentAppRequirements2.TargetNavigatorServer = BaseTargetNavigatorServer;
  ContentAppRequirements2.MediaPlaybackServer = BaseMediaPlaybackServer;
  ContentAppRequirements2.ContentLauncherServer = BaseContentLauncherServer;
  ContentAppRequirements2.AccountLoginServer = BaseAccountLoginServer;
  ContentAppRequirements2.ContentAppObserverBehavior = BaseContentAppObserverBehavior;
  ContentAppRequirements2.server = {
    mandatory: {
      KeypadInput: ContentAppRequirements2.KeypadInputServer,
      ApplicationLauncher: ContentAppRequirements2.ApplicationLauncherServer,
      ApplicationBasic: ContentAppRequirements2.ApplicationBasicServer
    },
    optional: {
      Binding: ContentAppRequirements2.BindingServer,
      Channel: ContentAppRequirements2.ChannelServer,
      TargetNavigator: ContentAppRequirements2.TargetNavigatorServer,
      MediaPlayback: ContentAppRequirements2.MediaPlaybackServer,
      ContentLauncher: ContentAppRequirements2.ContentLauncherServer,
      AccountLogin: ContentAppRequirements2.AccountLoginServer
    }
  };
  ContentAppRequirements2.client = { optional: { ContentAppObserver: ContentAppRequirements2.ContentAppObserverBehavior }, mandatory: {} };
})(ContentAppRequirements || (ContentAppRequirements = {}));
const ContentAppDeviceDefinition = MutableEndpoint({
  name: "ContentApp",
  deviceType: 36,
  deviceRevision: 2,
  requirements: ContentAppRequirements,
  behaviors: SupportedBehaviors(
    ContentAppRequirements.server.mandatory.KeypadInput,
    ContentAppRequirements.server.mandatory.ApplicationLauncher,
    ContentAppRequirements.server.mandatory.ApplicationBasic
  )
});
const ContentAppDevice = ContentAppDeviceDefinition;
export {
  ContentAppDevice,
  ContentAppDeviceDefinition,
  ContentAppRequirements
};
