// @project-chip/matter.js/dist/esm/endpoint/definitions/device/ContactSensorDevice.js
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { IdentifyServer as BaseIdentifyServer } from "../../../behavior/definitions/identify/IdentifyServer.js";
import {
  BooleanStateServer as BaseBooleanStateServer
} from "../../../behavior/definitions/boolean-state/BooleanStateServer.js";
import {
  BooleanStateConfigurationServer as BaseBooleanStateConfigurationServer
} from "../../../behavior/definitions/boolean-state-configuration/BooleanStateConfigurationServer.js";
import { MutableEndpoint } from "../../type/MutableEndpoint.js";
import { SupportedBehaviors } from "../../properties/SupportedBehaviors.js";
var ContactSensorRequirements;
((ContactSensorRequirements2) => {
  ContactSensorRequirements2.IdentifyServer = BaseIdentifyServer;
  ContactSensorRequirements2.BooleanStateServer = BaseBooleanStateServer;
  ContactSensorRequirements2.BooleanStateConfigurationServer = BaseBooleanStateConfigurationServer;
  ContactSensorRequirements2.server = {
    mandatory: { Identify: ContactSensorRequirements2.IdentifyServer, BooleanState: ContactSensorRequirements2.BooleanStateServer },
    optional: { BooleanStateConfiguration: ContactSensorRequirements2.BooleanStateConfigurationServer }
  };
})(ContactSensorRequirements || (ContactSensorRequirements = {}));
const ContactSensorDeviceDefinition = MutableEndpoint({
  name: "ContactSensor",
  deviceType: 21,
  deviceRevision: 2,
  requirements: ContactSensorRequirements,
  behaviors: SupportedBehaviors(
    ContactSensorRequirements.server.mandatory.Identify,
    ContactSensorRequirements.server.mandatory.BooleanState
  )
});
const ContactSensorDevice = ContactSensorDeviceDefinition;
export {
  ContactSensorDevice,
  ContactSensorDeviceDefinition,
  ContactSensorRequirements
};
