// @project-chip/matter.js/dist/esm/endpoint/definitions/device/DimmerSwitchDevice.js
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { IdentifyServer as BaseIdentifyServer } from "../../../behavior/definitions/identify/IdentifyServer.js";
import { IdentifyBehavior as BaseIdentifyBehavior } from "../../../behavior/definitions/identify/IdentifyBehavior.js";
import { OnOffBehavior as BaseOnOffBehavior } from "../../../behavior/definitions/on-off/OnOffBehavior.js";
import {
  LevelControlBehavior as BaseLevelControlBehavior
} from "../../../behavior/definitions/level-control/LevelControlBehavior.js";
import { GroupsBehavior as BaseGroupsBehavior } from "../../../behavior/definitions/groups/GroupsBehavior.js";
import {
  ScenesManagementBehavior as BaseScenesManagementBehavior
} from "../../../behavior/definitions/scenes-management/ScenesManagementBehavior.js";
import { MutableEndpoint } from "../../type/MutableEndpoint.js";
import { SupportedBehaviors } from "../../properties/SupportedBehaviors.js";
var DimmerSwitchRequirements;
((DimmerSwitchRequirements2) => {
  DimmerSwitchRequirements2.IdentifyServer = BaseIdentifyServer;
  DimmerSwitchRequirements2.IdentifyBehavior = BaseIdentifyBehavior;
  DimmerSwitchRequirements2.OnOffBehavior = BaseOnOffBehavior;
  DimmerSwitchRequirements2.LevelControlBehavior = BaseLevelControlBehavior;
  DimmerSwitchRequirements2.GroupsBehavior = BaseGroupsBehavior;
  DimmerSwitchRequirements2.ScenesManagementBehavior = BaseScenesManagementBehavior;
  DimmerSwitchRequirements2.server = { mandatory: { Identify: DimmerSwitchRequirements2.IdentifyServer } };
  DimmerSwitchRequirements2.client = {
    mandatory: { Identify: DimmerSwitchRequirements2.IdentifyBehavior, OnOff: DimmerSwitchRequirements2.OnOffBehavior, LevelControl: DimmerSwitchRequirements2.LevelControlBehavior },
    optional: { Groups: DimmerSwitchRequirements2.GroupsBehavior, ScenesManagement: DimmerSwitchRequirements2.ScenesManagementBehavior }
  };
})(DimmerSwitchRequirements || (DimmerSwitchRequirements = {}));
const DimmerSwitchDeviceDefinition = MutableEndpoint({
  name: "DimmerSwitch",
  deviceType: 260,
  deviceRevision: 3,
  requirements: DimmerSwitchRequirements,
  behaviors: SupportedBehaviors(DimmerSwitchRequirements.server.mandatory.Identify)
});
const DimmerSwitchDevice = DimmerSwitchDeviceDefinition;
export {
  DimmerSwitchDevice,
  DimmerSwitchDeviceDefinition,
  DimmerSwitchRequirements
};
