// @project-chip/matter.js/dist/esm/endpoint/definitions/device/DishwasherDevice.js
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  OperationalStateServer as BaseOperationalStateServer
} from "../../../behavior/definitions/operational-state/OperationalStateServer.js";
import { IdentifyServer as BaseIdentifyServer } from "../../../behavior/definitions/identify/IdentifyServer.js";
import { OnOffServer as BaseOnOffServer } from "../../../behavior/definitions/on-off/OnOffServer.js";
import {
  TemperatureControlServer as BaseTemperatureControlServer
} from "../../../behavior/definitions/temperature-control/TemperatureControlServer.js";
import {
  DishwasherModeServer as BaseDishwasherModeServer
} from "../../../behavior/definitions/dishwasher-mode/DishwasherModeServer.js";
import {
  DishwasherAlarmServer as BaseDishwasherAlarmServer
} from "../../../behavior/definitions/dishwasher-alarm/DishwasherAlarmServer.js";
import { MutableEndpoint } from "../../type/MutableEndpoint.js";
import { SupportedBehaviors } from "../../properties/SupportedBehaviors.js";
var DishwasherRequirements;
((DishwasherRequirements2) => {
  DishwasherRequirements2.OperationalStateServer = BaseOperationalStateServer;
  DishwasherRequirements2.IdentifyServer = BaseIdentifyServer;
  DishwasherRequirements2.OnOffServer = BaseOnOffServer.with("DeadFrontBehavior");
  DishwasherRequirements2.TemperatureControlServer = BaseTemperatureControlServer;
  DishwasherRequirements2.DishwasherModeServer = BaseDishwasherModeServer;
  DishwasherRequirements2.DishwasherAlarmServer = BaseDishwasherAlarmServer;
  DishwasherRequirements2.server = {
    mandatory: { OperationalState: DishwasherRequirements2.OperationalStateServer },
    optional: {
      Identify: DishwasherRequirements2.IdentifyServer,
      OnOff: DishwasherRequirements2.OnOffServer,
      TemperatureControl: DishwasherRequirements2.TemperatureControlServer,
      DishwasherMode: DishwasherRequirements2.DishwasherModeServer,
      DishwasherAlarm: DishwasherRequirements2.DishwasherAlarmServer
    }
  };
})(DishwasherRequirements || (DishwasherRequirements = {}));
const DishwasherDeviceDefinition = MutableEndpoint({
  name: "Dishwasher",
  deviceType: 117,
  deviceRevision: 1,
  requirements: DishwasherRequirements,
  behaviors: SupportedBehaviors(DishwasherRequirements.server.mandatory.OperationalState)
});
const DishwasherDevice = DishwasherDeviceDefinition;
export {
  DishwasherDevice,
  DishwasherDeviceDefinition,
  DishwasherRequirements
};
