// @project-chip/matter.js/dist/esm/endpoint/definitions/device/ExtendedColorLightDevice.js
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { IdentifyServer as BaseIdentifyServer } from "../../../behavior/definitions/identify/IdentifyServer.js";
import { GroupsServer as BaseGroupsServer } from "../../../behavior/definitions/groups/GroupsServer.js";
import { OnOffServer as BaseOnOffServer } from "../../../behavior/definitions/on-off/OnOffServer.js";
import {
  LevelControlServer as BaseLevelControlServer
} from "../../../behavior/definitions/level-control/LevelControlServer.js";
import {
  ColorControlServer as BaseColorControlServer
} from "../../../behavior/definitions/color-control/ColorControlServer.js";
import {
  ScenesManagementServer as BaseScenesManagementServer
} from "../../../behavior/definitions/scenes-management/ScenesManagementServer.js";
import {
  OccupancySensingBehavior as BaseOccupancySensingBehavior
} from "../../../behavior/definitions/occupancy-sensing/OccupancySensingBehavior.js";
import { MutableEndpoint } from "../../type/MutableEndpoint.js";
import { SupportedBehaviors } from "../../properties/SupportedBehaviors.js";
var ExtendedColorLightRequirements;
((ExtendedColorLightRequirements2) => {
  ExtendedColorLightRequirements2.IdentifyServer = BaseIdentifyServer.alter({ commands: { triggerEffect: { optional: false } } });
  ExtendedColorLightRequirements2.GroupsServer = BaseGroupsServer;
  ExtendedColorLightRequirements2.OnOffServer = BaseOnOffServer.with("Lighting");
  ExtendedColorLightRequirements2.LevelControlServer = BaseLevelControlServer.with("OnOff", "Lighting").alter({
    attributes: {
      currentLevel: { min: 1, max: 254 },
      minLevel: { default: 1, min: 1, max: 2 },
      maxLevel: { default: 254, min: 254, max: 255 }
    }
  });
  ExtendedColorLightRequirements2.ColorControlServer = BaseColorControlServer.with("Xy", "ColorTemperature").alter({ attributes: { remainingTime: { optional: false } } });
  ExtendedColorLightRequirements2.ScenesManagementServer = BaseScenesManagementServer;
  ExtendedColorLightRequirements2.OccupancySensingBehavior = BaseOccupancySensingBehavior;
  ExtendedColorLightRequirements2.server = {
    mandatory: {
      Identify: ExtendedColorLightRequirements2.IdentifyServer,
      Groups: ExtendedColorLightRequirements2.GroupsServer,
      OnOff: ExtendedColorLightRequirements2.OnOffServer,
      LevelControl: ExtendedColorLightRequirements2.LevelControlServer,
      ColorControl: ExtendedColorLightRequirements2.ColorControlServer
    },
    optional: { ScenesManagement: ExtendedColorLightRequirements2.ScenesManagementServer }
  };
  ExtendedColorLightRequirements2.client = { optional: { OccupancySensing: ExtendedColorLightRequirements2.OccupancySensingBehavior }, mandatory: {} };
})(ExtendedColorLightRequirements || (ExtendedColorLightRequirements = {}));
const ExtendedColorLightDeviceDefinition = MutableEndpoint({
  name: "ExtendedColorLight",
  deviceType: 269,
  deviceRevision: 4,
  requirements: ExtendedColorLightRequirements,
  behaviors: SupportedBehaviors(
    ExtendedColorLightRequirements.server.mandatory.Identify,
    ExtendedColorLightRequirements.server.mandatory.Groups,
    ExtendedColorLightRequirements.server.mandatory.OnOff,
    ExtendedColorLightRequirements.server.mandatory.LevelControl,
    ExtendedColorLightRequirements.server.mandatory.ColorControl
  )
});
const ExtendedColorLightDevice = ExtendedColorLightDeviceDefinition;
export {
  ExtendedColorLightDevice,
  ExtendedColorLightDeviceDefinition,
  ExtendedColorLightRequirements
};
