// @project-chip/matter.js/dist/esm/endpoint/definitions/device/LaundryWasherDevice.js
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  OperationalStateServer as BaseOperationalStateServer
} from "../../../behavior/definitions/operational-state/OperationalStateServer.js";
import { IdentifyServer as BaseIdentifyServer } from "../../../behavior/definitions/identify/IdentifyServer.js";
import {
  LaundryWasherModeServer as BaseLaundryWasherModeServer
} from "../../../behavior/definitions/laundry-washer-mode/LaundryWasherModeServer.js";
import { OnOffServer as BaseOnOffServer } from "../../../behavior/definitions/on-off/OnOffServer.js";
import {
  LaundryWasherControlsServer as BaseLaundryWasherControlsServer
} from "../../../behavior/definitions/laundry-washer-controls/LaundryWasherControlsServer.js";
import {
  TemperatureControlServer as BaseTemperatureControlServer
} from "../../../behavior/definitions/temperature-control/TemperatureControlServer.js";
import { MutableEndpoint } from "../../type/MutableEndpoint.js";
import { SupportedBehaviors } from "../../properties/SupportedBehaviors.js";
var LaundryWasherRequirements;
((LaundryWasherRequirements2) => {
  LaundryWasherRequirements2.OperationalStateServer = BaseOperationalStateServer;
  LaundryWasherRequirements2.IdentifyServer = BaseIdentifyServer;
  LaundryWasherRequirements2.LaundryWasherModeServer = BaseLaundryWasherModeServer;
  LaundryWasherRequirements2.OnOffServer = BaseOnOffServer.with("DeadFrontBehavior");
  LaundryWasherRequirements2.LaundryWasherControlsServer = BaseLaundryWasherControlsServer;
  LaundryWasherRequirements2.TemperatureControlServer = BaseTemperatureControlServer;
  LaundryWasherRequirements2.server = {
    mandatory: { OperationalState: LaundryWasherRequirements2.OperationalStateServer },
    optional: {
      Identify: LaundryWasherRequirements2.IdentifyServer,
      LaundryWasherMode: LaundryWasherRequirements2.LaundryWasherModeServer,
      OnOff: LaundryWasherRequirements2.OnOffServer,
      LaundryWasherControls: LaundryWasherRequirements2.LaundryWasherControlsServer,
      TemperatureControl: LaundryWasherRequirements2.TemperatureControlServer
    }
  };
})(LaundryWasherRequirements || (LaundryWasherRequirements = {}));
const LaundryWasherDeviceDefinition = MutableEndpoint({
  name: "LaundryWasher",
  deviceType: 115,
  deviceRevision: 1,
  requirements: LaundryWasherRequirements,
  behaviors: SupportedBehaviors(LaundryWasherRequirements.server.mandatory.OperationalState)
});
const LaundryWasherDevice = LaundryWasherDeviceDefinition;
export {
  LaundryWasherDevice,
  LaundryWasherDeviceDefinition,
  LaundryWasherRequirements
};
