// @project-chip/matter.js/dist/esm/endpoint/definitions/device/ModeSelectDevice.js
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ModeSelectServer as BaseModeSelectServer } from "../../../behavior/definitions/mode-select/ModeSelectServer.js";
import { MutableEndpoint } from "../../type/MutableEndpoint.js";
import { SupportedBehaviors } from "../../properties/SupportedBehaviors.js";
var ModeSelectRequirements;
((ModeSelectRequirements2) => {
  ModeSelectRequirements2.ModeSelectServer = BaseModeSelectServer;
  ModeSelectRequirements2.server = { mandatory: { ModeSelect: ModeSelectRequirements2.ModeSelectServer } };
})(ModeSelectRequirements || (ModeSelectRequirements = {}));
const ModeSelectDeviceDefinition = MutableEndpoint({
  name: "ModeSelect",
  deviceType: 39,
  deviceRevision: 1,
  requirements: ModeSelectRequirements,
  behaviors: SupportedBehaviors(ModeSelectRequirements.server.mandatory.ModeSelect)
});
const ModeSelectDevice = ModeSelectDeviceDefinition;
export {
  ModeSelectDevice,
  ModeSelectDeviceDefinition,
  ModeSelectRequirements
};
