// @project-chip/matter.js/dist/esm/endpoint/definitions/device/OccupancySensorDevice.js
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { IdentifyServer as BaseIdentifyServer } from "../../../behavior/definitions/identify/IdentifyServer.js";
import {
  OccupancySensingServer as BaseOccupancySensingServer
} from "../../../behavior/definitions/occupancy-sensing/OccupancySensingServer.js";
import { MutableEndpoint } from "../../type/MutableEndpoint.js";
import { SupportedBehaviors } from "../../properties/SupportedBehaviors.js";
var OccupancySensorRequirements;
((OccupancySensorRequirements2) => {
  OccupancySensorRequirements2.IdentifyServer = BaseIdentifyServer;
  OccupancySensorRequirements2.OccupancySensingServer = BaseOccupancySensingServer;
  OccupancySensorRequirements2.server = { mandatory: { Identify: OccupancySensorRequirements2.IdentifyServer, OccupancySensing: OccupancySensorRequirements2.OccupancySensingServer } };
})(OccupancySensorRequirements || (OccupancySensorRequirements = {}));
const OccupancySensorDeviceDefinition = MutableEndpoint({
  name: "OccupancySensor",
  deviceType: 263,
  deviceRevision: 3,
  requirements: OccupancySensorRequirements,
  behaviors: SupportedBehaviors(
    OccupancySensorRequirements.server.mandatory.Identify,
    OccupancySensorRequirements.server.mandatory.OccupancySensing
  )
});
const OccupancySensorDevice = OccupancySensorDeviceDefinition;
export {
  OccupancySensorDevice,
  OccupancySensorDeviceDefinition,
  OccupancySensorRequirements
};
