// @project-chip/matter.js/dist/esm/endpoint/definitions/device/OnOffLightSwitchDevice.js
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { IdentifyServer as BaseIdentifyServer } from "../../../behavior/definitions/identify/IdentifyServer.js";
import { IdentifyBehavior as BaseIdentifyBehavior } from "../../../behavior/definitions/identify/IdentifyBehavior.js";
import { OnOffBehavior as BaseOnOffBehavior } from "../../../behavior/definitions/on-off/OnOffBehavior.js";
import { GroupsBehavior as BaseGroupsBehavior } from "../../../behavior/definitions/groups/GroupsBehavior.js";
import {
  ScenesManagementBehavior as BaseScenesManagementBehavior
} from "../../../behavior/definitions/scenes-management/ScenesManagementBehavior.js";
import { MutableEndpoint } from "../../type/MutableEndpoint.js";
import { SupportedBehaviors } from "../../properties/SupportedBehaviors.js";
var OnOffLightSwitchRequirements;
((OnOffLightSwitchRequirements2) => {
  OnOffLightSwitchRequirements2.IdentifyServer = BaseIdentifyServer;
  OnOffLightSwitchRequirements2.IdentifyBehavior = BaseIdentifyBehavior;
  OnOffLightSwitchRequirements2.OnOffBehavior = BaseOnOffBehavior;
  OnOffLightSwitchRequirements2.GroupsBehavior = BaseGroupsBehavior;
  OnOffLightSwitchRequirements2.ScenesManagementBehavior = BaseScenesManagementBehavior;
  OnOffLightSwitchRequirements2.server = { mandatory: { Identify: OnOffLightSwitchRequirements2.IdentifyServer } };
  OnOffLightSwitchRequirements2.client = {
    mandatory: { Identify: OnOffLightSwitchRequirements2.IdentifyBehavior, OnOff: OnOffLightSwitchRequirements2.OnOffBehavior },
    optional: { Groups: OnOffLightSwitchRequirements2.GroupsBehavior, ScenesManagement: OnOffLightSwitchRequirements2.ScenesManagementBehavior }
  };
})(OnOffLightSwitchRequirements || (OnOffLightSwitchRequirements = {}));
const OnOffLightSwitchDeviceDefinition = MutableEndpoint({
  name: "OnOffLightSwitch",
  deviceType: 259,
  deviceRevision: 3,
  requirements: OnOffLightSwitchRequirements,
  behaviors: SupportedBehaviors(OnOffLightSwitchRequirements.server.mandatory.Identify)
});
const OnOffLightSwitchDevice = OnOffLightSwitchDeviceDefinition;
export {
  OnOffLightSwitchDevice,
  OnOffLightSwitchDeviceDefinition,
  OnOffLightSwitchRequirements
};
