// @project-chip/matter.js/dist/esm/endpoint/definitions/device/RefrigeratorDevice.js
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { IdentifyServer as BaseIdentifyServer } from "../../../behavior/definitions/identify/IdentifyServer.js";
import {
  RefrigeratorAndTemperatureControlledCabinetModeServer as BaseRefrigeratorAndTemperatureControlledCabinetModeServer
} from "../../../behavior/definitions/refrigerator-and-temperature-controlled-cabinet-mode/RefrigeratorAndTemperatureControlledCabinetModeServer.js";
import {
  RefrigeratorAlarmServer as BaseRefrigeratorAlarmServer
} from "../../../behavior/definitions/refrigerator-alarm/RefrigeratorAlarmServer.js";
import { MutableEndpoint } from "../../type/MutableEndpoint.js";
import { SupportedBehaviors } from "../../properties/SupportedBehaviors.js";
var RefrigeratorRequirements;
((RefrigeratorRequirements2) => {
  RefrigeratorRequirements2.IdentifyServer = BaseIdentifyServer;
  RefrigeratorRequirements2.RefrigeratorAndTemperatureControlledCabinetModeServer = BaseRefrigeratorAndTemperatureControlledCabinetModeServer;
  RefrigeratorRequirements2.RefrigeratorAlarmServer = BaseRefrigeratorAlarmServer;
  RefrigeratorRequirements2.server = {
    optional: {
      Identify: RefrigeratorRequirements2.IdentifyServer,
      RefrigeratorAndTemperatureControlledCabinetMode: RefrigeratorRequirements2.RefrigeratorAndTemperatureControlledCabinetModeServer,
      RefrigeratorAlarm: RefrigeratorRequirements2.RefrigeratorAlarmServer
    },
    mandatory: {}
  };
})(RefrigeratorRequirements || (RefrigeratorRequirements = {}));
const RefrigeratorDeviceDefinition = MutableEndpoint({
  name: "Refrigerator",
  deviceType: 112,
  deviceRevision: 1,
  requirements: RefrigeratorRequirements,
  behaviors: SupportedBehaviors()
});
const RefrigeratorDevice = RefrigeratorDeviceDefinition;
export {
  RefrigeratorDevice,
  RefrigeratorDeviceDefinition,
  RefrigeratorRequirements
};
