// @project-chip/matter.js/dist/esm/endpoint/definitions/device/RoboticVacuumCleanerDevice.js
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { IdentifyServer as BaseIdentifyServer } from "../../../behavior/definitions/identify/IdentifyServer.js";
import { RvcRunModeServer as BaseRvcRunModeServer } from "../../../behavior/definitions/rvc-run-mode/RvcRunModeServer.js";
import {
  RvcOperationalStateServer as BaseRvcOperationalStateServer
} from "../../../behavior/definitions/rvc-operational-state/RvcOperationalStateServer.js";
import {
  RvcCleanModeServer as BaseRvcCleanModeServer
} from "../../../behavior/definitions/rvc-clean-mode/RvcCleanModeServer.js";
import { MutableEndpoint } from "../../type/MutableEndpoint.js";
import { SupportedBehaviors } from "../../properties/SupportedBehaviors.js";
var RoboticVacuumCleanerRequirements;
((RoboticVacuumCleanerRequirements2) => {
  RoboticVacuumCleanerRequirements2.IdentifyServer = BaseIdentifyServer;
  RoboticVacuumCleanerRequirements2.RvcRunModeServer = BaseRvcRunModeServer;
  RoboticVacuumCleanerRequirements2.RvcOperationalStateServer = BaseRvcOperationalStateServer;
  RoboticVacuumCleanerRequirements2.RvcCleanModeServer = BaseRvcCleanModeServer;
  RoboticVacuumCleanerRequirements2.server = {
    mandatory: {
      Identify: RoboticVacuumCleanerRequirements2.IdentifyServer,
      RvcRunMode: RoboticVacuumCleanerRequirements2.RvcRunModeServer,
      RvcOperationalState: RoboticVacuumCleanerRequirements2.RvcOperationalStateServer
    },
    optional: { RvcCleanMode: RoboticVacuumCleanerRequirements2.RvcCleanModeServer }
  };
})(RoboticVacuumCleanerRequirements || (RoboticVacuumCleanerRequirements = {}));
const RoboticVacuumCleanerDeviceDefinition = MutableEndpoint({
  name: "RoboticVacuumCleaner",
  deviceType: 116,
  deviceRevision: 2,
  requirements: RoboticVacuumCleanerRequirements,
  behaviors: SupportedBehaviors(
    RoboticVacuumCleanerRequirements.server.mandatory.Identify,
    RoboticVacuumCleanerRequirements.server.mandatory.RvcRunMode,
    RoboticVacuumCleanerRequirements.server.mandatory.RvcOperationalState
  )
});
const RoboticVacuumCleanerDevice = RoboticVacuumCleanerDeviceDefinition;
export {
  RoboticVacuumCleanerDevice,
  RoboticVacuumCleanerDeviceDefinition,
  RoboticVacuumCleanerRequirements
};
