// @project-chip/matter.js/dist/esm/endpoint/definitions/device/SpeakerDevice.js
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { OnOffServer as BaseOnOffServer } from "../../../behavior/definitions/on-off/OnOffServer.js";
import {
  LevelControlServer as BaseLevelControlServer
} from "../../../behavior/definitions/level-control/LevelControlServer.js";
import { MutableEndpoint } from "../../type/MutableEndpoint.js";
import { SupportedBehaviors } from "../../properties/SupportedBehaviors.js";
var SpeakerRequirements;
((SpeakerRequirements2) => {
  SpeakerRequirements2.OnOffServer = BaseOnOffServer;
  SpeakerRequirements2.LevelControlServer = BaseLevelControlServer;
  SpeakerRequirements2.server = { mandatory: { OnOff: SpeakerRequirements2.OnOffServer, LevelControl: SpeakerRequirements2.LevelControlServer } };
})(SpeakerRequirements || (SpeakerRequirements = {}));
const SpeakerDeviceDefinition = MutableEndpoint({
  name: "Speaker",
  deviceType: 34,
  deviceRevision: 1,
  requirements: SpeakerRequirements,
  behaviors: SupportedBehaviors(
    SpeakerRequirements.server.mandatory.OnOff,
    SpeakerRequirements.server.mandatory.LevelControl
  )
});
const SpeakerDevice = SpeakerDeviceDefinition;
export {
  SpeakerDevice,
  SpeakerDeviceDefinition,
  SpeakerRequirements
};
