// @project-chip/matter.js/dist/esm/endpoint/definitions/device/TemperatureControlledCabinetDevice.js
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  TemperatureControlServer as BaseTemperatureControlServer
} from "../../../behavior/definitions/temperature-control/TemperatureControlServer.js";
import {
  TemperatureMeasurementServer as BaseTemperatureMeasurementServer
} from "../../../behavior/definitions/temperature-measurement/TemperatureMeasurementServer.js";
import {
  RefrigeratorAndTemperatureControlledCabinetModeServer as BaseRefrigeratorAndTemperatureControlledCabinetModeServer
} from "../../../behavior/definitions/refrigerator-and-temperature-controlled-cabinet-mode/RefrigeratorAndTemperatureControlledCabinetModeServer.js";
import { OvenModeServer as BaseOvenModeServer } from "../../../behavior/definitions/oven-mode/OvenModeServer.js";
import {
  OvenCavityOperationalStateServer as BaseOvenCavityOperationalStateServer
} from "../../../behavior/definitions/oven-cavity-operational-state/OvenCavityOperationalStateServer.js";
import { MutableEndpoint } from "../../type/MutableEndpoint.js";
import { SupportedBehaviors } from "../../properties/SupportedBehaviors.js";
var TemperatureControlledCabinetRequirements;
((TemperatureControlledCabinetRequirements2) => {
  TemperatureControlledCabinetRequirements2.TemperatureControlServer = BaseTemperatureControlServer;
  TemperatureControlledCabinetRequirements2.TemperatureMeasurementServer = BaseTemperatureMeasurementServer;
  TemperatureControlledCabinetRequirements2.RefrigeratorAndTemperatureControlledCabinetModeServer = BaseRefrigeratorAndTemperatureControlledCabinetModeServer;
  TemperatureControlledCabinetRequirements2.OvenModeServer = BaseOvenModeServer;
  TemperatureControlledCabinetRequirements2.OvenCavityOperationalStateServer = BaseOvenCavityOperationalStateServer;
  TemperatureControlledCabinetRequirements2.server = {
    mandatory: { TemperatureControl: TemperatureControlledCabinetRequirements2.TemperatureControlServer },
    optional: {
      TemperatureMeasurement: TemperatureControlledCabinetRequirements2.TemperatureMeasurementServer,
      RefrigeratorAndTemperatureControlledCabinetMode: TemperatureControlledCabinetRequirements2.RefrigeratorAndTemperatureControlledCabinetModeServer,
      OvenMode: TemperatureControlledCabinetRequirements2.OvenModeServer,
      OvenCavityOperationalState: TemperatureControlledCabinetRequirements2.OvenCavityOperationalStateServer
    }
  };
})(TemperatureControlledCabinetRequirements || (TemperatureControlledCabinetRequirements = {}));
const TemperatureControlledCabinetDeviceDefinition = MutableEndpoint({
  name: "TemperatureControlledCabinet",
  deviceType: 113,
  deviceRevision: 2,
  requirements: TemperatureControlledCabinetRequirements,
  behaviors: SupportedBehaviors()
});
const TemperatureControlledCabinetDevice = TemperatureControlledCabinetDeviceDefinition;
export {
  TemperatureControlledCabinetDevice,
  TemperatureControlledCabinetDeviceDefinition,
  TemperatureControlledCabinetRequirements
};
