// @project-chip/matter.js/dist/esm/endpoint/definitions/device/VideoRemoteControlDevice.js
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { OnOffBehavior as BaseOnOffBehavior } from "../../../behavior/definitions/on-off/OnOffBehavior.js";
import {
  MediaPlaybackBehavior as BaseMediaPlaybackBehavior
} from "../../../behavior/definitions/media-playback/MediaPlaybackBehavior.js";
import {
  KeypadInputBehavior as BaseKeypadInputBehavior
} from "../../../behavior/definitions/keypad-input/KeypadInputBehavior.js";
import {
  LevelControlBehavior as BaseLevelControlBehavior
} from "../../../behavior/definitions/level-control/LevelControlBehavior.js";
import { WakeOnLanBehavior as BaseWakeOnLanBehavior } from "../../../behavior/definitions/wake-on-lan/WakeOnLanBehavior.js";
import { ChannelBehavior as BaseChannelBehavior } from "../../../behavior/definitions/channel/ChannelBehavior.js";
import {
  TargetNavigatorBehavior as BaseTargetNavigatorBehavior
} from "../../../behavior/definitions/target-navigator/TargetNavigatorBehavior.js";
import {
  MediaInputBehavior as BaseMediaInputBehavior
} from "../../../behavior/definitions/media-input/MediaInputBehavior.js";
import { LowPowerBehavior as BaseLowPowerBehavior } from "../../../behavior/definitions/low-power/LowPowerBehavior.js";
import {
  ContentLauncherBehavior as BaseContentLauncherBehavior
} from "../../../behavior/definitions/content-launcher/ContentLauncherBehavior.js";
import {
  AudioOutputBehavior as BaseAudioOutputBehavior
} from "../../../behavior/definitions/audio-output/AudioOutputBehavior.js";
import {
  ApplicationLauncherBehavior as BaseApplicationLauncherBehavior
} from "../../../behavior/definitions/application-launcher/ApplicationLauncherBehavior.js";
import {
  AccountLoginBehavior as BaseAccountLoginBehavior
} from "../../../behavior/definitions/account-login/AccountLoginBehavior.js";
import {
  ContentControlBehavior as BaseContentControlBehavior
} from "../../../behavior/definitions/content-control/ContentControlBehavior.js";
import { MutableEndpoint } from "../../type/MutableEndpoint.js";
import { SupportedBehaviors } from "../../properties/SupportedBehaviors.js";
var VideoRemoteControlRequirements;
((VideoRemoteControlRequirements2) => {
  VideoRemoteControlRequirements2.OnOffBehavior = BaseOnOffBehavior;
  VideoRemoteControlRequirements2.MediaPlaybackBehavior = BaseMediaPlaybackBehavior;
  VideoRemoteControlRequirements2.KeypadInputBehavior = BaseKeypadInputBehavior;
  VideoRemoteControlRequirements2.LevelControlBehavior = BaseLevelControlBehavior;
  VideoRemoteControlRequirements2.WakeOnLanBehavior = BaseWakeOnLanBehavior;
  VideoRemoteControlRequirements2.ChannelBehavior = BaseChannelBehavior;
  VideoRemoteControlRequirements2.TargetNavigatorBehavior = BaseTargetNavigatorBehavior;
  VideoRemoteControlRequirements2.MediaInputBehavior = BaseMediaInputBehavior;
  VideoRemoteControlRequirements2.LowPowerBehavior = BaseLowPowerBehavior;
  VideoRemoteControlRequirements2.ContentLauncherBehavior = BaseContentLauncherBehavior;
  VideoRemoteControlRequirements2.AudioOutputBehavior = BaseAudioOutputBehavior;
  VideoRemoteControlRequirements2.ApplicationLauncherBehavior = BaseApplicationLauncherBehavior;
  VideoRemoteControlRequirements2.AccountLoginBehavior = BaseAccountLoginBehavior;
  VideoRemoteControlRequirements2.ContentControlBehavior = BaseContentControlBehavior;
  VideoRemoteControlRequirements2.client = {
    mandatory: { OnOff: VideoRemoteControlRequirements2.OnOffBehavior, MediaPlayback: VideoRemoteControlRequirements2.MediaPlaybackBehavior, KeypadInput: VideoRemoteControlRequirements2.KeypadInputBehavior },
    optional: {
      LevelControl: VideoRemoteControlRequirements2.LevelControlBehavior,
      WakeOnLan: VideoRemoteControlRequirements2.WakeOnLanBehavior,
      Channel: VideoRemoteControlRequirements2.ChannelBehavior,
      TargetNavigator: VideoRemoteControlRequirements2.TargetNavigatorBehavior,
      MediaInput: VideoRemoteControlRequirements2.MediaInputBehavior,
      LowPower: VideoRemoteControlRequirements2.LowPowerBehavior,
      ContentLauncher: VideoRemoteControlRequirements2.ContentLauncherBehavior,
      AudioOutput: VideoRemoteControlRequirements2.AudioOutputBehavior,
      ApplicationLauncher: VideoRemoteControlRequirements2.ApplicationLauncherBehavior,
      AccountLogin: VideoRemoteControlRequirements2.AccountLoginBehavior,
      ContentControl: VideoRemoteControlRequirements2.ContentControlBehavior
    }
  };
})(VideoRemoteControlRequirements || (VideoRemoteControlRequirements = {}));
const VideoRemoteControlDeviceDefinition = MutableEndpoint({
  name: "VideoRemoteControl",
  deviceType: 42,
  deviceRevision: 2,
  requirements: VideoRemoteControlRequirements,
  behaviors: SupportedBehaviors()
});
const VideoRemoteControlDevice = VideoRemoteControlDeviceDefinition;
export {
  VideoRemoteControlDevice,
  VideoRemoteControlDeviceDefinition,
  VideoRemoteControlRequirements
};
