// @project-chip/matter.js/dist/esm/endpoint/definitions/device/WaterLeakDetectorDevice.js
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { IdentifyServer as BaseIdentifyServer } from "../../../behavior/definitions/identify/IdentifyServer.js";
import {
  BooleanStateServer as BaseBooleanStateServer
} from "../../../behavior/definitions/boolean-state/BooleanStateServer.js";
import {
  BooleanStateConfigurationServer as BaseBooleanStateConfigurationServer
} from "../../../behavior/definitions/boolean-state-configuration/BooleanStateConfigurationServer.js";
import { MutableEndpoint } from "../../type/MutableEndpoint.js";
import { SupportedBehaviors } from "../../properties/SupportedBehaviors.js";
var WaterLeakDetectorRequirements;
((WaterLeakDetectorRequirements2) => {
  WaterLeakDetectorRequirements2.IdentifyServer = BaseIdentifyServer;
  WaterLeakDetectorRequirements2.BooleanStateServer = BaseBooleanStateServer.alter({ events: { stateChange: { optional: false } } });
  WaterLeakDetectorRequirements2.BooleanStateConfigurationServer = BaseBooleanStateConfigurationServer;
  WaterLeakDetectorRequirements2.server = {
    mandatory: { Identify: WaterLeakDetectorRequirements2.IdentifyServer, BooleanState: WaterLeakDetectorRequirements2.BooleanStateServer },
    optional: { BooleanStateConfiguration: WaterLeakDetectorRequirements2.BooleanStateConfigurationServer }
  };
})(WaterLeakDetectorRequirements || (WaterLeakDetectorRequirements = {}));
const WaterLeakDetectorDeviceDefinition = MutableEndpoint({
  name: "WaterLeakDetector",
  deviceType: 67,
  deviceRevision: 1,
  requirements: WaterLeakDetectorRequirements,
  behaviors: SupportedBehaviors(
    WaterLeakDetectorRequirements.server.mandatory.Identify,
    WaterLeakDetectorRequirements.server.mandatory.BooleanState
  )
});
const WaterLeakDetectorDevice = WaterLeakDetectorDeviceDefinition;
export {
  WaterLeakDetectorDevice,
  WaterLeakDetectorDeviceDefinition,
  WaterLeakDetectorRequirements
};
