// @project-chip/matter.js/dist/esm/endpoint/definitions/device/WindowCoveringControllerDevice.js
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { IdentifyServer as BaseIdentifyServer } from "../../../behavior/definitions/identify/IdentifyServer.js";
import {
  WindowCoveringBehavior as BaseWindowCoveringBehavior
} from "../../../behavior/definitions/window-covering/WindowCoveringBehavior.js";
import { IdentifyBehavior as BaseIdentifyBehavior } from "../../../behavior/definitions/identify/IdentifyBehavior.js";
import { GroupsBehavior as BaseGroupsBehavior } from "../../../behavior/definitions/groups/GroupsBehavior.js";
import {
  ScenesManagementBehavior as BaseScenesManagementBehavior
} from "../../../behavior/definitions/scenes-management/ScenesManagementBehavior.js";
import { MutableEndpoint } from "../../type/MutableEndpoint.js";
import { SupportedBehaviors } from "../../properties/SupportedBehaviors.js";
var WindowCoveringControllerRequirements;
((WindowCoveringControllerRequirements2) => {
  WindowCoveringControllerRequirements2.IdentifyServer = BaseIdentifyServer;
  WindowCoveringControllerRequirements2.WindowCoveringBehavior = BaseWindowCoveringBehavior;
  WindowCoveringControllerRequirements2.IdentifyBehavior = BaseIdentifyBehavior;
  WindowCoveringControllerRequirements2.GroupsBehavior = BaseGroupsBehavior;
  WindowCoveringControllerRequirements2.ScenesManagementBehavior = BaseScenesManagementBehavior;
  WindowCoveringControllerRequirements2.server = { optional: { Identify: WindowCoveringControllerRequirements2.IdentifyServer }, mandatory: {} };
  WindowCoveringControllerRequirements2.client = {
    mandatory: { WindowCovering: WindowCoveringControllerRequirements2.WindowCoveringBehavior },
    optional: { Identify: WindowCoveringControllerRequirements2.IdentifyBehavior, Groups: WindowCoveringControllerRequirements2.GroupsBehavior, ScenesManagement: WindowCoveringControllerRequirements2.ScenesManagementBehavior }
  };
})(WindowCoveringControllerRequirements || (WindowCoveringControllerRequirements = {}));
const WindowCoveringControllerDeviceDefinition = MutableEndpoint({
  name: "WindowCoveringController",
  deviceType: 515,
  deviceRevision: 3,
  requirements: WindowCoveringControllerRequirements,
  behaviors: SupportedBehaviors()
});
const WindowCoveringControllerDevice = WindowCoveringControllerDeviceDefinition;
export {
  WindowCoveringControllerDevice,
  WindowCoveringControllerDeviceDefinition,
  WindowCoveringControllerRequirements
};
