// @project-chip/matter.js/dist/esm/endpoint/definitions/device/WindowCoveringDevice.js
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { IdentifyServer as BaseIdentifyServer } from "../../../behavior/definitions/identify/IdentifyServer.js";
import {
  WindowCoveringServer as BaseWindowCoveringServer
} from "../../../behavior/definitions/window-covering/WindowCoveringServer.js";
import { GroupsServer as BaseGroupsServer } from "../../../behavior/definitions/groups/GroupsServer.js";
import {
  ScenesManagementServer as BaseScenesManagementServer
} from "../../../behavior/definitions/scenes-management/ScenesManagementServer.js";
import { MutableEndpoint } from "../../type/MutableEndpoint.js";
import { SupportedBehaviors } from "../../properties/SupportedBehaviors.js";
var WindowCoveringRequirements;
((WindowCoveringRequirements2) => {
  WindowCoveringRequirements2.IdentifyServer = BaseIdentifyServer;
  WindowCoveringRequirements2.WindowCoveringServer = BaseWindowCoveringServer;
  WindowCoveringRequirements2.GroupsServer = BaseGroupsServer;
  WindowCoveringRequirements2.ScenesManagementServer = BaseScenesManagementServer;
  WindowCoveringRequirements2.server = {
    mandatory: { Identify: WindowCoveringRequirements2.IdentifyServer, WindowCovering: WindowCoveringRequirements2.WindowCoveringServer },
    optional: { Groups: WindowCoveringRequirements2.GroupsServer, ScenesManagement: WindowCoveringRequirements2.ScenesManagementServer }
  };
})(WindowCoveringRequirements || (WindowCoveringRequirements = {}));
const WindowCoveringDeviceDefinition = MutableEndpoint({
  name: "WindowCovering",
  deviceType: 514,
  deviceRevision: 3,
  requirements: WindowCoveringRequirements,
  behaviors: SupportedBehaviors(WindowCoveringRequirements.server.mandatory.Identify)
});
const WindowCoveringDevice = WindowCoveringDeviceDefinition;
export {
  WindowCoveringDevice,
  WindowCoveringDeviceDefinition,
  WindowCoveringRequirements
};
