// @project-chip/matter.js/dist/cjs/cluster/definitions/DishwasherModeCluster.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DishwasherModeCluster_exports = {};
__export(DishwasherModeCluster_exports, {
  DishwasherMode: () => DishwasherMode,
  DishwasherModeCluster: () => DishwasherModeCluster
});
module.exports = __toCommonJS(DishwasherModeCluster_exports);
var import_MutableCluster = require("../mutation/MutableCluster.js");
var import_BitmapSchema = require("../../schema/BitmapSchema.js");
var import_Cluster = require("../Cluster.js");
var import_TlvArray = require("../../tlv/TlvArray.js");
var import_ModeBaseCluster = require("./ModeBaseCluster.js");
var import_TlvNumber = require("../../tlv/TlvNumber.js");
var import_ClusterRegistry = require("../ClusterRegistry.js");
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var DishwasherMode;
((DishwasherMode2) => {
  let Feature;
  ((Feature2) => {
    Feature2["OnOff"] = "OnOff";
  })(Feature = DishwasherMode2.Feature || (DishwasherMode2.Feature = {}));
  let ModeTag;
  ((ModeTag2) => {
    ModeTag2[ModeTag2["Normal"] = 16384] = "Normal";
    ModeTag2[ModeTag2["Heavy"] = 16385] = "Heavy";
    ModeTag2[ModeTag2["Light"] = 16386] = "Light";
  })(ModeTag = DishwasherMode2.ModeTag || (DishwasherMode2.ModeTag = {}));
  DishwasherMode2.Base = import_MutableCluster.MutableCluster.Component({
    id: 89,
    name: "DishwasherMode",
    revision: 2,
    features: {
      /**
       * OnOff
       *
       * This feature creates a dependency between an OnOff cluster instance and this cluster instance on the
       * same endpoint. See OnMode for more information.
       *
       * @see {@link MatterSpecification.v13.Cluster} § 1.10.4.1
       */
      onOff: (0, import_BitmapSchema.BitFlag)(0)
    },
    attributes: {
      /**
       * @see {@link MatterSpecification.v13.Cluster} § 8.3.5
       */
      supportedModes: (0, import_Cluster.FixedAttribute)(
        0,
        (0, import_TlvArray.TlvArray)(import_ModeBaseCluster.ModeBase.TlvModeOption, { minLength: 2, maxLength: 255 }),
        { default: [] }
      ),
      /**
       * @see {@link MatterSpecification.v13.Cluster} § 8.3.5
       */
      currentMode: (0, import_Cluster.Attribute)(1, import_TlvNumber.TlvUInt8, { scene: true, persistent: true }),
      /**
       * If this attribute is supported, the device SHOULD initially set this to one of the supported modes that
       * has the Normal tag associated with it. See the Mode Base cluster specification for full details about
       * the StartUpMode attribute.
       *
       * @see {@link MatterSpecification.v13.Cluster} § 8.3.5.1
       */
      startUpMode: (0, import_Cluster.WritableAttribute)(2, import_TlvNumber.TlvUInt8, { persistent: true }),
      /**
       * @see {@link MatterSpecification.v13.Cluster} § 8.3.5
       */
      onMode: (0, import_Cluster.WritableAttribute)(3, import_TlvNumber.TlvUInt8, { persistent: true })
    },
    commands: {
      /**
       * This command is used to change device modes.
       *
       * On receipt of this command the device shall respond with a ChangeToModeResponse command.
       *
       * @see {@link MatterSpecification.v13.Cluster} § 1.10.7.1
       */
      changeToMode: (0, import_Cluster.Command)(0, import_ModeBaseCluster.ModeBase.TlvChangeToModeRequest, 0, import_Cluster.TlvNoResponse)
    },
    /**
     * This metadata controls which DishwasherModeCluster elements matter.js activates for specific feature
     * combinations.
     */
    extensions: import_MutableCluster.MutableCluster.Extensions()
  });
  DishwasherMode2.ClusterInstance = (0, import_MutableCluster.MutableCluster)(DishwasherMode2.Base);
  DishwasherMode2.Cluster = DishwasherMode2.ClusterInstance;
  DishwasherMode2.Complete = DishwasherMode2.Cluster;
})(DishwasherMode || (DishwasherMode = {}));
const DishwasherModeCluster = DishwasherMode.Cluster;
import_ClusterRegistry.ClusterRegistry.register(DishwasherMode.Complete);
