// @project-chip/matter.js/dist/cjs/cluster/definitions/MicrowaveOvenModeCluster.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var MicrowaveOvenModeCluster_exports = {};
__export(MicrowaveOvenModeCluster_exports, {
  MicrowaveOvenMode: () => MicrowaveOvenMode,
  MicrowaveOvenModeCluster: () => MicrowaveOvenModeCluster
});
module.exports = __toCommonJS(MicrowaveOvenModeCluster_exports);
var import_MutableCluster = require("../mutation/MutableCluster.js");
var import_BitmapSchema = require("../../schema/BitmapSchema.js");
var import_Cluster = require("../Cluster.js");
var import_TlvArray = require("../../tlv/TlvArray.js");
var import_ModeBaseCluster = require("./ModeBaseCluster.js");
var import_TlvNumber = require("../../tlv/TlvNumber.js");
var import_ClusterRegistry = require("../ClusterRegistry.js");
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var MicrowaveOvenMode;
((MicrowaveOvenMode2) => {
  let Feature;
  ((Feature2) => {
    Feature2["OnOff"] = "OnOff";
  })(Feature = MicrowaveOvenMode2.Feature || (MicrowaveOvenMode2.Feature = {}));
  let ModeTag;
  ((ModeTag2) => {
    ModeTag2[ModeTag2["Normal"] = 16384] = "Normal";
    ModeTag2[ModeTag2["Defrost"] = 16385] = "Defrost";
  })(ModeTag = MicrowaveOvenMode2.ModeTag || (MicrowaveOvenMode2.ModeTag = {}));
  MicrowaveOvenMode2.Base = import_MutableCluster.MutableCluster.Component({
    id: 94,
    name: "MicrowaveOvenMode",
    revision: 1,
    features: {
      /**
       * OnOff
       *
       * This feature creates a dependency between an OnOff cluster instance and this cluster instance on the
       * same endpoint. See OnMode for more information.
       *
       * @see {@link MatterSpecification.v13.Cluster} ยง 1.10.4.1
       */
      onOff: (0, import_BitmapSchema.BitFlag)(0)
    },
    attributes: {
      /**
       * @see {@link MatterSpecification.v13.Cluster} ยง 8.12.4
       */
      supportedModes: (0, import_Cluster.FixedAttribute)(
        0,
        (0, import_TlvArray.TlvArray)(import_ModeBaseCluster.ModeBase.TlvModeOption, { minLength: 2, maxLength: 255 }),
        { default: [] }
      ),
      /**
       * @see {@link MatterSpecification.v13.Cluster} ยง 8.12.4
       */
      currentMode: (0, import_Cluster.Attribute)(1, import_TlvNumber.TlvUInt8, { scene: true, persistent: true })
    },
    /**
     * This metadata controls which MicrowaveOvenModeCluster elements matter.js activates for specific feature
     * combinations.
     */
    extensions: import_MutableCluster.MutableCluster.Extensions()
  });
  MicrowaveOvenMode2.ClusterInstance = (0, import_MutableCluster.MutableCluster)(MicrowaveOvenMode2.Base);
  MicrowaveOvenMode2.Cluster = MicrowaveOvenMode2.ClusterInstance;
  MicrowaveOvenMode2.Complete = MicrowaveOvenMode2.Cluster;
})(MicrowaveOvenMode || (MicrowaveOvenMode = {}));
const MicrowaveOvenModeCluster = MicrowaveOvenMode.Cluster;
import_ClusterRegistry.ClusterRegistry.register(MicrowaveOvenMode.Complete);
