// @project-chip/matter.js/dist/cjs/cluster/definitions/RefrigeratorAndTemperatureControlledCabinetModeCluster.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var RefrigeratorAndTemperatureControlledCabinetModeCluster_exports = {};
__export(RefrigeratorAndTemperatureControlledCabinetModeCluster_exports, {
  RefrigeratorAndTemperatureControlledCabinetMode: () => RefrigeratorAndTemperatureControlledCabinetMode,
  RefrigeratorAndTemperatureControlledCabinetModeCluster: () => RefrigeratorAndTemperatureControlledCabinetModeCluster
});
module.exports = __toCommonJS(RefrigeratorAndTemperatureControlledCabinetModeCluster_exports);
var import_MutableCluster = require("../mutation/MutableCluster.js");
var import_BitmapSchema = require("../../schema/BitmapSchema.js");
var import_Cluster = require("../Cluster.js");
var import_TlvArray = require("../../tlv/TlvArray.js");
var import_ModeBaseCluster = require("./ModeBaseCluster.js");
var import_TlvNumber = require("../../tlv/TlvNumber.js");
var import_ClusterRegistry = require("../ClusterRegistry.js");
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var RefrigeratorAndTemperatureControlledCabinetMode;
((RefrigeratorAndTemperatureControlledCabinetMode2) => {
  let Feature;
  ((Feature2) => {
    Feature2["OnOff"] = "OnOff";
  })(Feature = RefrigeratorAndTemperatureControlledCabinetMode2.Feature || (RefrigeratorAndTemperatureControlledCabinetMode2.Feature = {}));
  let ModeTag;
  ((ModeTag2) => {
    ModeTag2[ModeTag2["RapidCool"] = 16384] = "RapidCool";
    ModeTag2[ModeTag2["RapidFreeze"] = 16385] = "RapidFreeze";
  })(ModeTag = RefrigeratorAndTemperatureControlledCabinetMode2.ModeTag || (RefrigeratorAndTemperatureControlledCabinetMode2.ModeTag = {}));
  RefrigeratorAndTemperatureControlledCabinetMode2.Base = import_MutableCluster.MutableCluster.Component({
    id: 82,
    name: "RefrigeratorAndTemperatureControlledCabinetMode",
    revision: 2,
    features: {
      /**
       * OnOff
       *
       * This feature creates a dependency between an OnOff cluster instance and this cluster instance on the
       * same endpoint. See OnMode for more information.
       *
       * @see {@link MatterSpecification.v13.Cluster} § 1.10.4.1
       */
      onOff: (0, import_BitmapSchema.BitFlag)(0)
    },
    attributes: {
      /**
       * @see {@link MatterSpecification.v13.Cluster} § 8.7.5
       */
      supportedModes: (0, import_Cluster.FixedAttribute)(
        0,
        (0, import_TlvArray.TlvArray)(import_ModeBaseCluster.ModeBase.TlvModeOption, { minLength: 2, maxLength: 255 }),
        { default: [] }
      ),
      /**
       * @see {@link MatterSpecification.v13.Cluster} § 8.7.5
       */
      currentMode: (0, import_Cluster.Attribute)(1, import_TlvNumber.TlvUInt8, { scene: true, persistent: true }),
      /**
       * @see {@link MatterSpecification.v13.Cluster} § 8.7.5
       */
      startUpMode: (0, import_Cluster.WritableAttribute)(2, import_TlvNumber.TlvUInt8, { persistent: true }),
      /**
       * @see {@link MatterSpecification.v13.Cluster} § 8.7.5
       */
      onMode: (0, import_Cluster.WritableAttribute)(3, import_TlvNumber.TlvUInt8, { persistent: true })
    },
    commands: {
      /**
       * This command is used to change device modes.
       *
       * On receipt of this command the device shall respond with a ChangeToModeResponse command.
       *
       * @see {@link MatterSpecification.v13.Cluster} § 1.10.7.1
       */
      changeToMode: (0, import_Cluster.Command)(0, import_ModeBaseCluster.ModeBase.TlvChangeToModeRequest, 0, import_Cluster.TlvNoResponse)
    },
    /**
     * This metadata controls which RefrigeratorAndTemperatureControlledCabinetModeCluster elements matter.js
     * activates for specific feature combinations.
     */
    extensions: import_MutableCluster.MutableCluster.Extensions()
  });
  RefrigeratorAndTemperatureControlledCabinetMode2.ClusterInstance = (0, import_MutableCluster.MutableCluster)(RefrigeratorAndTemperatureControlledCabinetMode2.Base);
  RefrigeratorAndTemperatureControlledCabinetMode2.Cluster = RefrigeratorAndTemperatureControlledCabinetMode2.ClusterInstance;
  RefrigeratorAndTemperatureControlledCabinetMode2.Complete = RefrigeratorAndTemperatureControlledCabinetMode2.Cluster;
})(RefrigeratorAndTemperatureControlledCabinetMode || (RefrigeratorAndTemperatureControlledCabinetMode = {}));
const RefrigeratorAndTemperatureControlledCabinetModeCluster = RefrigeratorAndTemperatureControlledCabinetMode.Cluster;
import_ClusterRegistry.ClusterRegistry.register(RefrigeratorAndTemperatureControlledCabinetMode.Complete);
