// @project-chip/matter.js/dist/esm/endpoint/definitions/system/DeviceEnergyManagementEndpoint.js
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  DeviceEnergyManagementServer as BaseDeviceEnergyManagementServer
} from "../../../behavior/definitions/device-energy-management/DeviceEnergyManagementServer.js";
import {
  DeviceEnergyManagementModeServer as BaseDeviceEnergyManagementModeServer
} from "../../../behavior/definitions/device-energy-management-mode/DeviceEnergyManagementModeServer.js";
import { MutableEndpoint } from "../../type/MutableEndpoint.js";
import { DeviceClasses } from "../../../device/DeviceTypes.js";
import { SupportedBehaviors } from "../../properties/SupportedBehaviors.js";
var DeviceEnergyManagementRequirements;
((DeviceEnergyManagementRequirements2) => {
  DeviceEnergyManagementRequirements2.DeviceEnergyManagementServer = BaseDeviceEnergyManagementServer;
  DeviceEnergyManagementRequirements2.DeviceEnergyManagementModeServer = BaseDeviceEnergyManagementModeServer;
  DeviceEnergyManagementRequirements2.server = {
    optional: {
      DeviceEnergyManagement: DeviceEnergyManagementRequirements2.DeviceEnergyManagementServer,
      DeviceEnergyManagementMode: DeviceEnergyManagementRequirements2.DeviceEnergyManagementModeServer
    },
    mandatory: {}
  };
})(DeviceEnergyManagementRequirements || (DeviceEnergyManagementRequirements = {}));
const DeviceEnergyManagementEndpointDefinition = MutableEndpoint({
  name: "DeviceEnergyManagement",
  deviceType: 1293,
  deviceRevision: 1,
  deviceClass: DeviceClasses.Utility,
  requirements: DeviceEnergyManagementRequirements,
  behaviors: SupportedBehaviors()
});
const DeviceEnergyManagementEndpoint = DeviceEnergyManagementEndpointDefinition;
export {
  DeviceEnergyManagementEndpoint,
  DeviceEnergyManagementEndpointDefinition,
  DeviceEnergyManagementRequirements
};
