// @project-chip/matter.js/dist/esm/endpoint/definitions/system/ElectricalSensorEndpoint.js
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  PowerTopologyServer as BasePowerTopologyServer
} from "../../../behavior/definitions/power-topology/PowerTopologyServer.js";
import {
  ElectricalPowerMeasurementServer as BaseElectricalPowerMeasurementServer
} from "../../../behavior/definitions/electrical-power-measurement/ElectricalPowerMeasurementServer.js";
import {
  ElectricalEnergyMeasurementServer as BaseElectricalEnergyMeasurementServer
} from "../../../behavior/definitions/electrical-energy-measurement/ElectricalEnergyMeasurementServer.js";
import { MutableEndpoint } from "../../type/MutableEndpoint.js";
import { DeviceClasses } from "../../../device/DeviceTypes.js";
import { SupportedBehaviors } from "../../properties/SupportedBehaviors.js";
var ElectricalSensorRequirements;
((ElectricalSensorRequirements2) => {
  ElectricalSensorRequirements2.PowerTopologyServer = BasePowerTopologyServer;
  ElectricalSensorRequirements2.ElectricalPowerMeasurementServer = BaseElectricalPowerMeasurementServer;
  ElectricalSensorRequirements2.ElectricalEnergyMeasurementServer = BaseElectricalEnergyMeasurementServer;
  ElectricalSensorRequirements2.server = {
    mandatory: { PowerTopology: ElectricalSensorRequirements2.PowerTopologyServer },
    optional: {
      ElectricalPowerMeasurement: ElectricalSensorRequirements2.ElectricalPowerMeasurementServer,
      ElectricalEnergyMeasurement: ElectricalSensorRequirements2.ElectricalEnergyMeasurementServer
    }
  };
})(ElectricalSensorRequirements || (ElectricalSensorRequirements = {}));
const ElectricalSensorEndpointDefinition = MutableEndpoint({
  name: "ElectricalSensor",
  deviceType: 1296,
  deviceRevision: 1,
  deviceClass: DeviceClasses.Utility,
  requirements: ElectricalSensorRequirements,
  behaviors: SupportedBehaviors()
});
const ElectricalSensorEndpoint = ElectricalSensorEndpointDefinition;
export {
  ElectricalSensorEndpoint,
  ElectricalSensorEndpointDefinition,
  ElectricalSensorRequirements
};
