// @project-chip/matter.js/dist/esm/endpoint/definitions/system/OtaProviderEndpoint.js
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  OtaSoftwareUpdateProviderServer as BaseOtaSoftwareUpdateProviderServer
} from "../../../behavior/definitions/ota-software-update-provider/OtaSoftwareUpdateProviderServer.js";
import {
  OtaSoftwareUpdateRequestorBehavior as BaseOtaSoftwareUpdateRequestorBehavior
} from "../../../behavior/definitions/ota-software-update-requestor/OtaSoftwareUpdateRequestorBehavior.js";
import { MutableEndpoint } from "../../type/MutableEndpoint.js";
import { DeviceClasses } from "../../../device/DeviceTypes.js";
import { SupportedBehaviors } from "../../properties/SupportedBehaviors.js";
var OtaProviderRequirements;
((OtaProviderRequirements2) => {
  OtaProviderRequirements2.OtaSoftwareUpdateProviderServer = BaseOtaSoftwareUpdateProviderServer;
  OtaProviderRequirements2.OtaSoftwareUpdateRequestorBehavior = BaseOtaSoftwareUpdateRequestorBehavior;
  OtaProviderRequirements2.server = { mandatory: { OtaSoftwareUpdateProvider: OtaProviderRequirements2.OtaSoftwareUpdateProviderServer } };
  OtaProviderRequirements2.client = {
    optional: { OtaSoftwareUpdateRequestor: OtaProviderRequirements2.OtaSoftwareUpdateRequestorBehavior },
    mandatory: {}
  };
})(OtaProviderRequirements || (OtaProviderRequirements = {}));
const OtaProviderEndpointDefinition = MutableEndpoint({
  name: "OtaProvider",
  deviceType: 20,
  deviceRevision: 1,
  deviceClass: DeviceClasses.Utility,
  requirements: OtaProviderRequirements,
  behaviors: SupportedBehaviors(OtaProviderRequirements.server.mandatory.OtaSoftwareUpdateProvider)
});
const OtaProviderEndpoint = OtaProviderEndpointDefinition;
export {
  OtaProviderEndpoint,
  OtaProviderEndpointDefinition,
  OtaProviderRequirements
};
