// @project-chip/matter.js/dist/esm/endpoint/definitions/system/PowerSourceEndpoint.js
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  PowerSourceServer as BasePowerSourceServer
} from "../../../behavior/definitions/power-source/PowerSourceServer.js";
import { MutableEndpoint } from "../../type/MutableEndpoint.js";
import { DeviceClasses } from "../../../device/DeviceTypes.js";
import { SupportedBehaviors } from "../../properties/SupportedBehaviors.js";
var PowerSourceRequirements;
((PowerSourceRequirements2) => {
  PowerSourceRequirements2.PowerSourceServer = BasePowerSourceServer;
  PowerSourceRequirements2.server = { mandatory: { PowerSource: PowerSourceRequirements2.PowerSourceServer } };
})(PowerSourceRequirements || (PowerSourceRequirements = {}));
const PowerSourceEndpointDefinition = MutableEndpoint({
  name: "PowerSource",
  deviceType: 17,
  deviceRevision: 1,
  deviceClass: DeviceClasses.Utility,
  requirements: PowerSourceRequirements,
  behaviors: SupportedBehaviors(PowerSourceRequirements.server.mandatory.PowerSource)
});
const PowerSourceEndpoint = PowerSourceEndpointDefinition;
export {
  PowerSourceEndpoint,
  PowerSourceEndpointDefinition,
  PowerSourceRequirements
};
