// @project-chip/matter.js/dist/cjs/net/fake/SimulatedNetwork.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SimulatedNetwork_exports = {};
__export(SimulatedNetwork_exports, {
  FAKE_INTERFACE_NAME: () => FAKE_INTERFACE_NAME,
  SimulatedNetwork: () => SimulatedNetwork
});
module.exports = __toCommonJS(SimulatedNetwork_exports);
var import_Logger = require("../../log/Logger.js");
var import_Singleton = require("../../util/Singleton.js");
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const logger = import_Logger.Logger.get("SimulatedNetwork");
const FAKE_INTERFACE_NAME = "fakeInterface";
class SimulatedNetwork {
  static get = (0, import_Singleton.singleton)(() => new SimulatedNetwork());
  listenersMap = /* @__PURE__ */ new Map();
  onUdpData(host, port, listener) {
    const ipPort = `${host ?? "*"}:${port}`;
    let listeners = this.listenersMap.get(ipPort);
    if (listeners === void 0) {
      listeners = new Array();
      this.listenersMap.set(ipPort, listeners);
    }
    listeners.push(listener);
    return {
      close: async () => this.offUdpData(host, port, listener)
    };
  }
  offUdpData(host, port, listenerToRemove) {
    const ipPort = `${host ?? "*"}:${port}`;
    const listeners = this.listenersMap.get(ipPort);
    if (listeners === void 0) return;
    const newListeners = listeners.filter((listener) => listener !== listenerToRemove);
    if (newListeners.length === 0) {
      this.listenersMap.delete(ipPort);
      return;
    }
    this.listenersMap.set(ipPort, newListeners);
  }
  sendUdp(localAddress, localPort, remoteAddress, remotePort, data) {
    [`${remoteAddress}:${remotePort}`, `*:${remotePort}`].forEach(
      (ipPort) => this.listenersMap.get(ipPort)?.forEach((listener) => {
        try {
          listener(FAKE_INTERFACE_NAME, localAddress, localPort, data);
        } catch (error) {
          logger.error(error);
        }
      })
    );
  }
}
