// @project-chip/matter.js/dist/cjs/behavior/definitions/general-commissioning/GeneralCommissioningServer.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var GeneralCommissioningServer_exports = {};
__export(GeneralCommissioningServer_exports, {
  GeneralCommissioningServer: () => GeneralCommissioningServer
});
module.exports = __toCommonJS(GeneralCommissioningServer_exports);
var import_AdministratorCommissioningCluster = require("../../../cluster/definitions/AdministratorCommissioningCluster.js");
var import_GeneralCommissioningCluster = require("../../../cluster/definitions/GeneralCommissioningCluster.js");
var import_MatterError = require("../../../common/MatterError.js");
var import_Logger = require("../../../log/Logger.js");
var import_SecureSession = require("../../../session/SecureSession.js");
var import_AdministratorCommissioningServer = require("../administrator-commissioning/AdministratorCommissioningServer.js");
var import_BasicInformationServer = require("../basic-information/BasicInformationServer.js");
var import_GeneralCommissioningBehavior = require("./GeneralCommissioningBehavior.js");
var import_ServerNodeFailsafeContext = require("./ServerNodeFailsafeContext.js");
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const SuccessResponse = { errorCode: import_GeneralCommissioningCluster.GeneralCommissioning.CommissioningError.Ok, debugText: "" };
const logger = import_Logger.Logger.get("GeneralCommissioningClusterHandler");
class GeneralCommissioningServer extends import_GeneralCommissioningBehavior.GeneralCommissioningBehavior {
  initialize() {
    const bci = this.state.basicCommissioningInfo;
    if (bci.failSafeExpiryLengthSeconds === void 0) {
      bci.failSafeExpiryLengthSeconds = 60;
    }
    if (bci.maxCumulativeFailsafeSeconds === void 0) {
      bci.maxCumulativeFailsafeSeconds = 900;
    }
    this.state.breadcrumb = 0;
  }
  async armFailSafe({ breadcrumb, expiryLengthSeconds }) {
    (0, import_SecureSession.assertSecureSession)(this.session, "armFailSafe can only be called on a secure session");
    const device = this.session.context;
    try {
      if (!device.isFailsafeArmed() && this.agent.get(import_AdministratorCommissioningServer.AdministratorCommissioningServer).state.windowStatus !== import_AdministratorCommissioningCluster.AdministratorCommissioning.CommissioningWindowStatus.WindowNotOpen && !this.session.isPase) {
        throw new import_MatterError.MatterFlowError("Failed to arm failsafe using CASE while commissioning window is opened.");
      }
      if (device.isFailsafeArmed()) {
        await device.failsafeContext.extend(this.session.fabric, expiryLengthSeconds);
      } else {
        if (expiryLengthSeconds === 0) return SuccessResponse;
        await device.beginTimed(
          new import_ServerNodeFailsafeContext.ServerNodeFailsafeContext(this.endpoint, {
            fabrics: device.fabricManager,
            sessions: device.sessionManager,
            expiryLengthSeconds,
            maxCumulativeFailsafeSeconds: this.state.basicCommissioningInfo.maxCumulativeFailsafeSeconds,
            associatedFabric: this.session.fabric
          })
        );
      }
      if (device.isFailsafeArmed()) {
        this.state.breadcrumb = breadcrumb;
      }
    } catch (error) {
      import_MatterError.MatterFlowError.accept(error);
      logger.debug(`Error while arming failSafe timer`, error);
      return {
        errorCode: import_GeneralCommissioningCluster.GeneralCommissioning.CommissioningError.BusyWithOtherAdmin,
        debugText: error.message
      };
    }
    return SuccessResponse;
  }
  async setRegulatoryConfig({
    breadcrumb,
    newRegulatoryConfig,
    countryCode
  }) {
    const locationCapabilityValue = this.state.locationCapability;
    const basicInformation = this.agent.get(import_BasicInformationServer.BasicInformationServer);
    const currentLocationCountryCode = basicInformation.state.location;
    if (currentLocationCountryCode !== countryCode) {
      if (this.state.allowCountryCodeChange === false && countryCode !== "XX") {
        return {
          errorCode: import_GeneralCommissioningCluster.GeneralCommissioning.CommissioningError.ValueOutsideRange,
          debugText: `Country code change not allowed: ${countryCode}`
        };
      }
      if (this.state.countryCodeWhitelist !== void 0 && !this.state.countryCodeWhitelist.includes(countryCode)) {
        return {
          errorCode: import_GeneralCommissioningCluster.GeneralCommissioning.CommissioningError.ValueOutsideRange,
          debugText: `Country code change not allowed: ${countryCode}`
        };
      }
      if (countryCode !== "XX") {
        basicInformation.state.location = countryCode;
      }
    }
    let validValues;
    switch (locationCapabilityValue) {
      case import_GeneralCommissioningCluster.GeneralCommissioning.RegulatoryLocationType.Outdoor:
        validValues = [import_GeneralCommissioningCluster.GeneralCommissioning.RegulatoryLocationType.Outdoor];
        break;
      case import_GeneralCommissioningCluster.GeneralCommissioning.RegulatoryLocationType.Indoor:
        validValues = [import_GeneralCommissioningCluster.GeneralCommissioning.RegulatoryLocationType.Indoor];
        break;
      case import_GeneralCommissioningCluster.GeneralCommissioning.RegulatoryLocationType.IndoorOutdoor:
        validValues = [
          import_GeneralCommissioningCluster.GeneralCommissioning.RegulatoryLocationType.Indoor,
          import_GeneralCommissioningCluster.GeneralCommissioning.RegulatoryLocationType.Outdoor,
          import_GeneralCommissioningCluster.GeneralCommissioning.RegulatoryLocationType.IndoorOutdoor
        ];
        break;
      default:
        return {
          errorCode: import_GeneralCommissioningCluster.GeneralCommissioning.CommissioningError.ValueOutsideRange,
          debugText: `Invalid regulatory location: ${newRegulatoryConfig === import_GeneralCommissioningCluster.GeneralCommissioning.RegulatoryLocationType.Indoor ? "Indoor" : "Outdoor"}`
        };
    }
    if (!validValues.includes(newRegulatoryConfig)) {
      return {
        errorCode: import_GeneralCommissioningCluster.GeneralCommissioning.CommissioningError.ValueOutsideRange,
        debugText: `Invalid regulatory location: ${newRegulatoryConfig === import_GeneralCommissioningCluster.GeneralCommissioning.RegulatoryLocationType.Indoor ? "Indoor" : "Outdoor"}`
      };
    }
    this.asAdmin(() => {
      this.state.regulatoryConfig = newRegulatoryConfig;
    });
    this.state.breadcrumb = breadcrumb;
    return SuccessResponse;
  }
  async commissioningComplete() {
    if (this.session.isPase) {
      return {
        errorCode: import_GeneralCommissioningCluster.GeneralCommissioning.CommissioningError.InvalidAuthentication,
        debugText: "Command not executed over CASE session."
      };
    }
    const fabric = this.session.associatedFabric;
    const device = this.session.context;
    if (!device.isFailsafeArmed()) {
      return { errorCode: import_GeneralCommissioningCluster.GeneralCommissioning.CommissioningError.NoFailSafe, debugText: "FailSafe not armed." };
    }
    const failsafeContext = device.failsafeContext;
    (0, import_SecureSession.assertSecureSession)(this.session, "commissioningComplete can only be called on a secure session");
    const timedFabric = failsafeContext.associatedFabric?.fabricIndex;
    if (fabric.fabricIndex !== timedFabric) {
      return {
        errorCode: import_GeneralCommissioningCluster.GeneralCommissioning.CommissioningError.InvalidAuthentication,
        debugText: `Associated fabric ${fabric.fabricIndex} does not match the one from the failsafe context ${timedFabric}.`
      };
    }
    await failsafeContext.completeCommission();
    this.state.breadcrumb = BigInt(0);
    logger.info(`Commissioning completed on fabric #${fabric.fabricId} as node #${fabric.nodeId}.`);
    return SuccessResponse;
  }
}
((GeneralCommissioningServer2) => {
  class State extends import_GeneralCommissioningBehavior.GeneralCommissioningBehavior.State {
    /**
     * Set to false to prevent the controller from changing the country code during commissioning.
     */
    allowCountryCodeChange = true;
    // Default true if not set
    /**
     * Set to an array of two-letter country codes to limit the countries the controller may assign.
     */
    countryCodeWhitelist = void 0;
  }
  GeneralCommissioningServer2.State = State;
})(GeneralCommissioningServer || (GeneralCommissioningServer = {}));
