// @project-chip/matter.js/dist/cjs/endpoint/definitions/system/ElectricalSensorEndpoint.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ElectricalSensorEndpoint_exports = {};
__export(ElectricalSensorEndpoint_exports, {
  ElectricalSensorEndpoint: () => ElectricalSensorEndpoint,
  ElectricalSensorEndpointDefinition: () => ElectricalSensorEndpointDefinition,
  ElectricalSensorRequirements: () => ElectricalSensorRequirements
});
module.exports = __toCommonJS(ElectricalSensorEndpoint_exports);
var import_PowerTopologyServer = require("../../../behavior/definitions/power-topology/PowerTopologyServer.js");
var import_ElectricalPowerMeasurementServer = require("../../../behavior/definitions/electrical-power-measurement/ElectricalPowerMeasurementServer.js");
var import_ElectricalEnergyMeasurementServer = require("../../../behavior/definitions/electrical-energy-measurement/ElectricalEnergyMeasurementServer.js");
var import_MutableEndpoint = require("../../type/MutableEndpoint.js");
var import_DeviceTypes = require("../../../device/DeviceTypes.js");
var import_SupportedBehaviors = require("../../properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var ElectricalSensorRequirements;
((ElectricalSensorRequirements2) => {
  ElectricalSensorRequirements2.PowerTopologyServer = import_PowerTopologyServer.PowerTopologyServer;
  ElectricalSensorRequirements2.ElectricalPowerMeasurementServer = import_ElectricalPowerMeasurementServer.ElectricalPowerMeasurementServer;
  ElectricalSensorRequirements2.ElectricalEnergyMeasurementServer = import_ElectricalEnergyMeasurementServer.ElectricalEnergyMeasurementServer;
  ElectricalSensorRequirements2.server = {
    mandatory: { PowerTopology: ElectricalSensorRequirements2.PowerTopologyServer },
    optional: {
      ElectricalPowerMeasurement: ElectricalSensorRequirements2.ElectricalPowerMeasurementServer,
      ElectricalEnergyMeasurement: ElectricalSensorRequirements2.ElectricalEnergyMeasurementServer
    }
  };
})(ElectricalSensorRequirements || (ElectricalSensorRequirements = {}));
const ElectricalSensorEndpointDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "ElectricalSensor",
  deviceType: 1296,
  deviceRevision: 1,
  deviceClass: import_DeviceTypes.DeviceClasses.Utility,
  requirements: ElectricalSensorRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)()
});
const ElectricalSensorEndpoint = ElectricalSensorEndpointDefinition;
