// @project-chip/matter.js/dist/cjs/behavior/definitions/operational-credentials/OperationalCredentialsTypes.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var OperationalCredentialsTypes_exports = {};
__export(OperationalCredentialsTypes_exports, {
  TlvAttestation: () => TlvAttestation,
  TlvCertSigningRequest: () => TlvCertSigningRequest
});
module.exports = __toCommonJS(OperationalCredentialsTypes_exports);
var import_TlvNumber = require("../../../tlv/TlvNumber.js");
var import_TlvObject = require("../../../tlv/TlvObject.js");
var import_TlvString = require("../../../tlv/TlvString.js");
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const TlvAttestation = (0, import_TlvObject.TlvObject)({
  declaration: (0, import_TlvObject.TlvField)(1, import_TlvString.TlvByteString),
  attestationNonce: (0, import_TlvObject.TlvField)(2, import_TlvString.TlvByteString.bound({ length: 32 })),
  timestamp: (0, import_TlvObject.TlvField)(3, import_TlvNumber.TlvUInt32),
  // TODO: check actual max length in specs
  firmwareInfo: (0, import_TlvObject.TlvOptionalField)(4, import_TlvString.TlvByteString)
});
const TlvCertSigningRequest = (0, import_TlvObject.TlvObject)({
  certSigningRequest: (0, import_TlvObject.TlvField)(1, import_TlvString.TlvByteString),
  csrNonce: (0, import_TlvObject.TlvField)(2, import_TlvString.TlvByteString.bound({ length: 32 })),
  vendorReserved1: (0, import_TlvObject.TlvOptionalField)(3, import_TlvString.TlvByteString),
  vendorReserved2: (0, import_TlvObject.TlvOptionalField)(4, import_TlvString.TlvByteString),
  vendorReserved3: (0, import_TlvObject.TlvOptionalField)(5, import_TlvString.TlvByteString)
});
