// @project-chip/matter.js/dist/cjs/endpoint/properties/Parts.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Parts_exports = {};
__export(Parts_exports, {
  Parts: () => Parts
});
module.exports = __toCommonJS(Parts_exports);
var import_Lifecycle = require("../../common/Lifecycle.js");
var import_MatterError = require("../../common/MatterError.js");
var import_Logger = require("../../log/Logger.js");
var import_IdentityService = require("../../node/server/IdentityService.js");
var import_Construction = require("../../util/Construction.js");
var import_Set = require("../../util/Set.js");
var import_Agent = require("../Agent.js");
var import_Endpoint = require("../Endpoint.js");
var import_errors = require("../errors.js");
var import_EndpointLifecycle = require("./EndpointLifecycle.js");
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const logger = import_Logger.Logger.get("Parts");
class Parts {
  #bubbleChange;
  #children = new import_Set.BasicSet();
  #endpoint;
  constructor(endpoint) {
    this.#endpoint = endpoint;
    const lifecycle = this.#endpoint.lifecycle;
    this.#bubbleChange = (type, endpoint2) => lifecycle.bubble(type, endpoint2);
  }
  add(child) {
    const endpoint = this.#endpointFor(child);
    if (this.has(endpoint)) {
      return;
    }
    if (endpoint.lifecycle.hasId) {
      this.assertIdAvailable(endpoint.id, endpoint);
    }
    if (this.#endpoint.lifecycle.isReady) {
      this.#validateInsertion(endpoint, endpoint);
    }
    this.#children.add(endpoint);
    endpoint.owner = this.#endpoint;
    endpoint.lifecycle.changed.on((type, endpoint2) => this.#bubbleChange(type, endpoint2));
    endpoint.lifecycle.destroyed.once(() => {
      this.delete(endpoint);
    });
    if (this.#endpoint.lifecycle.isPartsReady) {
      if (!endpoint.construction.isErrorHandled) {
        endpoint.construction.onError((error) => logger.error(`Error initializing ${this.#endpoint}:`, error));
      }
      endpoint.construction.start();
    }
    return endpoint;
  }
  delete(child) {
    const endpoint = this.#endpointFor(child);
    if (!this.#children.delete(this.#endpointFor(child))) {
      return false;
    }
    const childLifeCycle = endpoint.lifecycle;
    childLifeCycle.changed.off(this.#bubbleChange);
    this.#children.delete(endpoint);
    return true;
  }
  clear() {
    this.#children.clear();
  }
  get(id) {
    for (const child of this.#children) {
      if (child.maybeId === id) {
        return child;
      }
    }
  }
  require(id) {
    const part = this.get(id);
    if (part === void 0) {
      throw new import_errors.PartNotFoundError(`Endpoint ${this.#endpoint} has no part ${id}`);
    }
    return part;
  }
  #get(id) {
    for (const child of this.#children) {
      if (child.maybeId === id) {
        return child;
      }
    }
  }
  has(identity) {
    if (typeof identity === "string") {
      for (const child of this.#children) {
        if (child.maybeId === identity) {
          return true;
        }
      }
      return false;
    }
    return this.#children.has(this.#endpointFor(identity));
  }
  indexOf(child) {
    const endpoint = this.#endpointFor(child);
    let index = 0;
    for (const other of this.#children) {
      if (endpoint === other) {
        return index;
      }
      index++;
    }
    return -1;
  }
  get added() {
    return this.#children.added;
  }
  get deleted() {
    return this.#children.deleted;
  }
  get size() {
    return this.#children.size;
  }
  [Symbol.iterator]() {
    return this.#children[Symbol.iterator]();
  }
  /**
   * Confirm availability of an ID amongst the endpoint's children.
   */
  assertIdAvailable(id, endpoint) {
    const other = this.#get(id);
    if (other && other !== endpoint) {
      throw new import_IdentityService.IdentityConflictError(`${other} is already defined; endpoint IDs must be unique within parent`);
    }
  }
  /**
   * Initialize all uninitialized Parts.
   *
   * Invoked automatically by the owner after behaviors initialize.
   */
  initialize() {
    if (!this.#endpoint.lifecycle.isReady) {
      throw new import_MatterError.ImplementationError(`Cannot initialize parts because endpoint is not ready`);
    }
    const onPartsReady = () => this.#endpoint.lifecycle.change(import_EndpointLifecycle.EndpointLifecycle.Change.PartsReady);
    if (!this.size) {
      onPartsReady();
      return;
    }
    for (const endpoint of this) {
      const isInactive = endpoint.construction.status === import_Lifecycle.Lifecycle.Status.Inactive;
      if (isInactive || endpoint.construction.status === import_Lifecycle.Lifecycle.Status.Initializing) {
        endpoint.construction.onError((error) => {
          logger.error(`Error initializing ${endpoint}:`, error);
        });
        if (isInactive) {
          endpoint.construction.start();
        }
      }
    }
    const onPartsCrashed = (errored) => {
      const essential = [...errored].filter((endpoint) => endpoint.lifecycle.isEssential);
      if (essential.length) {
        return new import_errors.EndpointPartsError(essential);
      }
    };
    let promise = import_Construction.Construction.all(this, onPartsCrashed);
    if (promise) {
      promise = promise.then(onPartsReady);
    } else {
      onPartsReady();
    }
    return promise;
  }
  /**
   * Soft-reset all parts.  Invoked automatically by the owner on reset.
   */
  async reset() {
    for (const endpoint of this) {
      await endpoint.reset();
    }
  }
  /**
   * Destroy all parts.  Invoked automatically by the owner on destroy.
   */
  async close() {
    for (const endpoint of this) {
      await endpoint.close();
    }
  }
  #validateInsertion(forefather, endpoint, usedNumbers) {
    if (endpoint.lifecycle.hasNumber) {
      this.#endpoint.env.get(import_IdentityService.IdentityService).assertNumberAvailable(endpoint.number, endpoint);
      if (usedNumbers?.has(endpoint.number)) {
        throw new import_IdentityService.IdentityConflictError(
          `Cannot add endpoint ${forefather} because descendents have conflicting definitions for endpoint number ${endpoint.number}`
        );
      }
    }
    if (!endpoint.hasParts) {
      return;
    }
    const children = endpoint.parts;
    if (!children.size) {
      return;
    }
    if (!usedNumbers) {
      usedNumbers = /* @__PURE__ */ new Set();
    }
    if (endpoint.lifecycle.hasNumber) {
      usedNumbers.add(endpoint.number);
    }
    for (const child of children) {
      this.#validateInsertion(forefather, child, usedNumbers);
    }
  }
  #endpointFor(child) {
    if (child instanceof import_Agent.Agent) {
      child = child.endpoint;
    }
    if (!(child instanceof import_Endpoint.Endpoint)) {
      if (child.type) {
        child.owner = this.#endpoint;
      } else {
        child = {
          type: child,
          owner: this.#endpoint
        };
      }
    }
    return import_Endpoint.Endpoint.partFor(child);
  }
}
