// @project-chip/matter.js/dist/cjs/log/Level.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Level_exports = {};
__export(Level_exports, {
  Level: () => Level,
  levelFromString: () => levelFromString
});
module.exports = __toCommonJS(Level_exports);
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
function levelFromString(level) {
  switch (level.toLowerCase()) {
    case "debug":
      return 0 /* DEBUG */;
    case "info":
      return 1 /* INFO */;
    case "notice":
      return 2 /* NOTICE */;
    case "warn":
      return 3 /* WARN */;
    case "error":
      return 4 /* ERROR */;
    case "fatal":
      return 5 /* FATAL */;
    default:
      return 1 /* INFO */;
  }
}
var Level = /* @__PURE__ */ ((Level2) => {
  Level2[Level2["DEBUG"] = 0] = "DEBUG";
  Level2[Level2["INFO"] = 1] = "INFO";
  Level2[Level2["NOTICE"] = 2] = "NOTICE";
  Level2[Level2["WARN"] = 3] = "WARN";
  Level2[Level2["ERROR"] = 4] = "ERROR";
  Level2[Level2["FATAL"] = 5] = "FATAL";
  return Level2;
})(Level || {});
