// @project-chip/matter.js/dist/cjs/endpoint/definitions/device/ColorDimmerSwitchDevice.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ColorDimmerSwitchDevice_exports = {};
__export(ColorDimmerSwitchDevice_exports, {
  ColorDimmerSwitchDevice: () => ColorDimmerSwitchDevice,
  ColorDimmerSwitchDeviceDefinition: () => ColorDimmerSwitchDeviceDefinition,
  ColorDimmerSwitchRequirements: () => ColorDimmerSwitchRequirements
});
module.exports = __toCommonJS(ColorDimmerSwitchDevice_exports);
var import_IdentifyServer = require("../../../behavior/definitions/identify/IdentifyServer.js");
var import_IdentifyBehavior = require("../../../behavior/definitions/identify/IdentifyBehavior.js");
var import_OnOffBehavior = require("../../../behavior/definitions/on-off/OnOffBehavior.js");
var import_LevelControlBehavior = require("../../../behavior/definitions/level-control/LevelControlBehavior.js");
var import_ColorControlBehavior = require("../../../behavior/definitions/color-control/ColorControlBehavior.js");
var import_GroupsBehavior = require("../../../behavior/definitions/groups/GroupsBehavior.js");
var import_ScenesManagementBehavior = require("../../../behavior/definitions/scenes-management/ScenesManagementBehavior.js");
var import_MutableEndpoint = require("../../type/MutableEndpoint.js");
var import_SupportedBehaviors = require("../../properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var ColorDimmerSwitchRequirements;
((ColorDimmerSwitchRequirements2) => {
  ColorDimmerSwitchRequirements2.IdentifyServer = import_IdentifyServer.IdentifyServer;
  ColorDimmerSwitchRequirements2.IdentifyBehavior = import_IdentifyBehavior.IdentifyBehavior;
  ColorDimmerSwitchRequirements2.OnOffBehavior = import_OnOffBehavior.OnOffBehavior;
  ColorDimmerSwitchRequirements2.LevelControlBehavior = import_LevelControlBehavior.LevelControlBehavior;
  ColorDimmerSwitchRequirements2.ColorControlBehavior = import_ColorControlBehavior.ColorControlBehavior;
  ColorDimmerSwitchRequirements2.GroupsBehavior = import_GroupsBehavior.GroupsBehavior;
  ColorDimmerSwitchRequirements2.ScenesManagementBehavior = import_ScenesManagementBehavior.ScenesManagementBehavior;
  ColorDimmerSwitchRequirements2.server = { mandatory: { Identify: ColorDimmerSwitchRequirements2.IdentifyServer } };
  ColorDimmerSwitchRequirements2.client = {
    mandatory: {
      Identify: ColorDimmerSwitchRequirements2.IdentifyBehavior,
      OnOff: ColorDimmerSwitchRequirements2.OnOffBehavior,
      LevelControl: ColorDimmerSwitchRequirements2.LevelControlBehavior,
      ColorControl: ColorDimmerSwitchRequirements2.ColorControlBehavior
    },
    optional: { Groups: ColorDimmerSwitchRequirements2.GroupsBehavior, ScenesManagement: ColorDimmerSwitchRequirements2.ScenesManagementBehavior }
  };
})(ColorDimmerSwitchRequirements || (ColorDimmerSwitchRequirements = {}));
const ColorDimmerSwitchDeviceDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "ColorDimmerSwitch",
  deviceType: 261,
  deviceRevision: 3,
  requirements: ColorDimmerSwitchRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)(ColorDimmerSwitchRequirements.server.mandatory.Identify)
});
const ColorDimmerSwitchDevice = ColorDimmerSwitchDeviceDefinition;
