// @project-chip/matter.js/dist/cjs/endpoint/definitions/device/ContentAppDevice.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ContentAppDevice_exports = {};
__export(ContentAppDevice_exports, {
  ContentAppDevice: () => ContentAppDevice,
  ContentAppDeviceDefinition: () => ContentAppDeviceDefinition,
  ContentAppRequirements: () => ContentAppRequirements
});
module.exports = __toCommonJS(ContentAppDevice_exports);
var import_KeypadInputServer = require("../../../behavior/definitions/keypad-input/KeypadInputServer.js");
var import_ApplicationLauncherServer = require("../../../behavior/definitions/application-launcher/ApplicationLauncherServer.js");
var import_ApplicationBasicServer = require("../../../behavior/definitions/application-basic/ApplicationBasicServer.js");
var import_BindingServer = require("../../../behavior/definitions/binding/BindingServer.js");
var import_ChannelServer = require("../../../behavior/definitions/channel/ChannelServer.js");
var import_TargetNavigatorServer = require("../../../behavior/definitions/target-navigator/TargetNavigatorServer.js");
var import_MediaPlaybackServer = require("../../../behavior/definitions/media-playback/MediaPlaybackServer.js");
var import_ContentLauncherServer = require("../../../behavior/definitions/content-launcher/ContentLauncherServer.js");
var import_AccountLoginServer = require("../../../behavior/definitions/account-login/AccountLoginServer.js");
var import_ContentAppObserverBehavior = require("../../../behavior/definitions/content-app-observer/ContentAppObserverBehavior.js");
var import_MutableEndpoint = require("../../type/MutableEndpoint.js");
var import_SupportedBehaviors = require("../../properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var ContentAppRequirements;
((ContentAppRequirements2) => {
  ContentAppRequirements2.KeypadInputServer = import_KeypadInputServer.KeypadInputServer;
  ContentAppRequirements2.ApplicationLauncherServer = import_ApplicationLauncherServer.ApplicationLauncherServer;
  ContentAppRequirements2.ApplicationBasicServer = import_ApplicationBasicServer.ApplicationBasicServer;
  ContentAppRequirements2.BindingServer = import_BindingServer.BindingServer;
  ContentAppRequirements2.ChannelServer = import_ChannelServer.ChannelServer;
  ContentAppRequirements2.TargetNavigatorServer = import_TargetNavigatorServer.TargetNavigatorServer;
  ContentAppRequirements2.MediaPlaybackServer = import_MediaPlaybackServer.MediaPlaybackServer;
  ContentAppRequirements2.ContentLauncherServer = import_ContentLauncherServer.ContentLauncherServer;
  ContentAppRequirements2.AccountLoginServer = import_AccountLoginServer.AccountLoginServer;
  ContentAppRequirements2.ContentAppObserverBehavior = import_ContentAppObserverBehavior.ContentAppObserverBehavior;
  ContentAppRequirements2.server = {
    mandatory: {
      KeypadInput: ContentAppRequirements2.KeypadInputServer,
      ApplicationLauncher: ContentAppRequirements2.ApplicationLauncherServer,
      ApplicationBasic: ContentAppRequirements2.ApplicationBasicServer
    },
    optional: {
      Binding: ContentAppRequirements2.BindingServer,
      Channel: ContentAppRequirements2.ChannelServer,
      TargetNavigator: ContentAppRequirements2.TargetNavigatorServer,
      MediaPlayback: ContentAppRequirements2.MediaPlaybackServer,
      ContentLauncher: ContentAppRequirements2.ContentLauncherServer,
      AccountLogin: ContentAppRequirements2.AccountLoginServer
    }
  };
  ContentAppRequirements2.client = { optional: { ContentAppObserver: ContentAppRequirements2.ContentAppObserverBehavior }, mandatory: {} };
})(ContentAppRequirements || (ContentAppRequirements = {}));
const ContentAppDeviceDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "ContentApp",
  deviceType: 36,
  deviceRevision: 2,
  requirements: ContentAppRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)(
    ContentAppRequirements.server.mandatory.KeypadInput,
    ContentAppRequirements.server.mandatory.ApplicationLauncher,
    ContentAppRequirements.server.mandatory.ApplicationBasic
  )
});
const ContentAppDevice = ContentAppDeviceDefinition;
