// @project-chip/matter.js/dist/cjs/endpoint/definitions/device/ControlBridgeDevice.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ControlBridgeDevice_exports = {};
__export(ControlBridgeDevice_exports, {
  ControlBridgeDevice: () => ControlBridgeDevice,
  ControlBridgeDeviceDefinition: () => ControlBridgeDeviceDefinition,
  ControlBridgeRequirements: () => ControlBridgeRequirements
});
module.exports = __toCommonJS(ControlBridgeDevice_exports);
var import_IdentifyServer = require("../../../behavior/definitions/identify/IdentifyServer.js");
var import_IdentifyBehavior = require("../../../behavior/definitions/identify/IdentifyBehavior.js");
var import_GroupsBehavior = require("../../../behavior/definitions/groups/GroupsBehavior.js");
var import_OnOffBehavior = require("../../../behavior/definitions/on-off/OnOffBehavior.js");
var import_LevelControlBehavior = require("../../../behavior/definitions/level-control/LevelControlBehavior.js");
var import_ColorControlBehavior = require("../../../behavior/definitions/color-control/ColorControlBehavior.js");
var import_ScenesManagementBehavior = require("../../../behavior/definitions/scenes-management/ScenesManagementBehavior.js");
var import_IlluminanceMeasurementBehavior = require("../../../behavior/definitions/illuminance-measurement/IlluminanceMeasurementBehavior.js");
var import_OccupancySensingBehavior = require("../../../behavior/definitions/occupancy-sensing/OccupancySensingBehavior.js");
var import_MutableEndpoint = require("../../type/MutableEndpoint.js");
var import_SupportedBehaviors = require("../../properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var ControlBridgeRequirements;
((ControlBridgeRequirements2) => {
  ControlBridgeRequirements2.IdentifyServer = import_IdentifyServer.IdentifyServer;
  ControlBridgeRequirements2.IdentifyBehavior = import_IdentifyBehavior.IdentifyBehavior;
  ControlBridgeRequirements2.GroupsBehavior = import_GroupsBehavior.GroupsBehavior;
  ControlBridgeRequirements2.OnOffBehavior = import_OnOffBehavior.OnOffBehavior;
  ControlBridgeRequirements2.LevelControlBehavior = import_LevelControlBehavior.LevelControlBehavior;
  ControlBridgeRequirements2.ColorControlBehavior = import_ColorControlBehavior.ColorControlBehavior;
  ControlBridgeRequirements2.ScenesManagementBehavior = import_ScenesManagementBehavior.ScenesManagementBehavior;
  ControlBridgeRequirements2.IlluminanceMeasurementBehavior = import_IlluminanceMeasurementBehavior.IlluminanceMeasurementBehavior;
  ControlBridgeRequirements2.OccupancySensingBehavior = import_OccupancySensingBehavior.OccupancySensingBehavior;
  ControlBridgeRequirements2.server = { mandatory: { Identify: ControlBridgeRequirements2.IdentifyServer } };
  ControlBridgeRequirements2.client = {
    mandatory: {
      Identify: ControlBridgeRequirements2.IdentifyBehavior,
      Groups: ControlBridgeRequirements2.GroupsBehavior,
      OnOff: ControlBridgeRequirements2.OnOffBehavior,
      LevelControl: ControlBridgeRequirements2.LevelControlBehavior,
      ColorControl: ControlBridgeRequirements2.ColorControlBehavior
    },
    optional: {
      ScenesManagement: ControlBridgeRequirements2.ScenesManagementBehavior,
      IlluminanceMeasurement: ControlBridgeRequirements2.IlluminanceMeasurementBehavior,
      OccupancySensing: ControlBridgeRequirements2.OccupancySensingBehavior
    }
  };
})(ControlBridgeRequirements || (ControlBridgeRequirements = {}));
const ControlBridgeDeviceDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "ControlBridge",
  deviceType: 2112,
  deviceRevision: 3,
  requirements: ControlBridgeRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)(ControlBridgeRequirements.server.mandatory.Identify)
});
const ControlBridgeDevice = ControlBridgeDeviceDefinition;
