// @project-chip/matter.js/dist/cjs/endpoint/definitions/device/ExtendedColorLightDevice.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ExtendedColorLightDevice_exports = {};
__export(ExtendedColorLightDevice_exports, {
  ExtendedColorLightDevice: () => ExtendedColorLightDevice,
  ExtendedColorLightDeviceDefinition: () => ExtendedColorLightDeviceDefinition,
  ExtendedColorLightRequirements: () => ExtendedColorLightRequirements
});
module.exports = __toCommonJS(ExtendedColorLightDevice_exports);
var import_IdentifyServer = require("../../../behavior/definitions/identify/IdentifyServer.js");
var import_GroupsServer = require("../../../behavior/definitions/groups/GroupsServer.js");
var import_OnOffServer = require("../../../behavior/definitions/on-off/OnOffServer.js");
var import_LevelControlServer = require("../../../behavior/definitions/level-control/LevelControlServer.js");
var import_ColorControlServer = require("../../../behavior/definitions/color-control/ColorControlServer.js");
var import_ScenesManagementServer = require("../../../behavior/definitions/scenes-management/ScenesManagementServer.js");
var import_OccupancySensingBehavior = require("../../../behavior/definitions/occupancy-sensing/OccupancySensingBehavior.js");
var import_MutableEndpoint = require("../../type/MutableEndpoint.js");
var import_SupportedBehaviors = require("../../properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var ExtendedColorLightRequirements;
((ExtendedColorLightRequirements2) => {
  ExtendedColorLightRequirements2.IdentifyServer = import_IdentifyServer.IdentifyServer.alter({ commands: { triggerEffect: { optional: false } } });
  ExtendedColorLightRequirements2.GroupsServer = import_GroupsServer.GroupsServer;
  ExtendedColorLightRequirements2.OnOffServer = import_OnOffServer.OnOffServer.with("Lighting");
  ExtendedColorLightRequirements2.LevelControlServer = import_LevelControlServer.LevelControlServer.with("OnOff", "Lighting").alter({
    attributes: {
      currentLevel: { min: 1, max: 254 },
      minLevel: { default: 1, min: 1, max: 2 },
      maxLevel: { default: 254, min: 254, max: 255 }
    }
  });
  ExtendedColorLightRequirements2.ColorControlServer = import_ColorControlServer.ColorControlServer.with("Xy", "ColorTemperature").alter({ attributes: { remainingTime: { optional: false } } });
  ExtendedColorLightRequirements2.ScenesManagementServer = import_ScenesManagementServer.ScenesManagementServer;
  ExtendedColorLightRequirements2.OccupancySensingBehavior = import_OccupancySensingBehavior.OccupancySensingBehavior;
  ExtendedColorLightRequirements2.server = {
    mandatory: {
      Identify: ExtendedColorLightRequirements2.IdentifyServer,
      Groups: ExtendedColorLightRequirements2.GroupsServer,
      OnOff: ExtendedColorLightRequirements2.OnOffServer,
      LevelControl: ExtendedColorLightRequirements2.LevelControlServer,
      ColorControl: ExtendedColorLightRequirements2.ColorControlServer
    },
    optional: { ScenesManagement: ExtendedColorLightRequirements2.ScenesManagementServer }
  };
  ExtendedColorLightRequirements2.client = { optional: { OccupancySensing: ExtendedColorLightRequirements2.OccupancySensingBehavior }, mandatory: {} };
})(ExtendedColorLightRequirements || (ExtendedColorLightRequirements = {}));
const ExtendedColorLightDeviceDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "ExtendedColorLight",
  deviceType: 269,
  deviceRevision: 4,
  requirements: ExtendedColorLightRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)(
    ExtendedColorLightRequirements.server.mandatory.Identify,
    ExtendedColorLightRequirements.server.mandatory.Groups,
    ExtendedColorLightRequirements.server.mandatory.OnOff,
    ExtendedColorLightRequirements.server.mandatory.LevelControl,
    ExtendedColorLightRequirements.server.mandatory.ColorControl
  )
});
const ExtendedColorLightDevice = ExtendedColorLightDeviceDefinition;
