// @project-chip/matter.js/dist/cjs/endpoint/definitions/device/LightSensorDevice.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var LightSensorDevice_exports = {};
__export(LightSensorDevice_exports, {
  LightSensorDevice: () => LightSensorDevice,
  LightSensorDeviceDefinition: () => LightSensorDeviceDefinition,
  LightSensorRequirements: () => LightSensorRequirements
});
module.exports = __toCommonJS(LightSensorDevice_exports);
var import_IdentifyServer = require("../../../behavior/definitions/identify/IdentifyServer.js");
var import_IlluminanceMeasurementServer = require("../../../behavior/definitions/illuminance-measurement/IlluminanceMeasurementServer.js");
var import_MutableEndpoint = require("../../type/MutableEndpoint.js");
var import_SupportedBehaviors = require("../../properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var LightSensorRequirements;
((LightSensorRequirements2) => {
  LightSensorRequirements2.IdentifyServer = import_IdentifyServer.IdentifyServer;
  LightSensorRequirements2.IlluminanceMeasurementServer = import_IlluminanceMeasurementServer.IlluminanceMeasurementServer;
  LightSensorRequirements2.server = { mandatory: { Identify: LightSensorRequirements2.IdentifyServer, IlluminanceMeasurement: LightSensorRequirements2.IlluminanceMeasurementServer } };
})(LightSensorRequirements || (LightSensorRequirements = {}));
const LightSensorDeviceDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "LightSensor",
  deviceType: 262,
  deviceRevision: 3,
  requirements: LightSensorRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)(
    LightSensorRequirements.server.mandatory.Identify,
    LightSensorRequirements.server.mandatory.IlluminanceMeasurement
  )
});
const LightSensorDevice = LightSensorDeviceDefinition;
