// @project-chip/matter.js/dist/cjs/endpoint/definitions/device/OnOffSensorDevice.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var OnOffSensorDevice_exports = {};
__export(OnOffSensorDevice_exports, {
  OnOffSensorDevice: () => OnOffSensorDevice,
  OnOffSensorDeviceDefinition: () => OnOffSensorDeviceDefinition,
  OnOffSensorRequirements: () => OnOffSensorRequirements
});
module.exports = __toCommonJS(OnOffSensorDevice_exports);
var import_IdentifyServer = require("../../../behavior/definitions/identify/IdentifyServer.js");
var import_IdentifyBehavior = require("../../../behavior/definitions/identify/IdentifyBehavior.js");
var import_OnOffBehavior = require("../../../behavior/definitions/on-off/OnOffBehavior.js");
var import_GroupsBehavior = require("../../../behavior/definitions/groups/GroupsBehavior.js");
var import_ScenesManagementBehavior = require("../../../behavior/definitions/scenes-management/ScenesManagementBehavior.js");
var import_LevelControlBehavior = require("../../../behavior/definitions/level-control/LevelControlBehavior.js");
var import_ColorControlBehavior = require("../../../behavior/definitions/color-control/ColorControlBehavior.js");
var import_MutableEndpoint = require("../../type/MutableEndpoint.js");
var import_SupportedBehaviors = require("../../properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var OnOffSensorRequirements;
((OnOffSensorRequirements2) => {
  OnOffSensorRequirements2.IdentifyServer = import_IdentifyServer.IdentifyServer;
  OnOffSensorRequirements2.IdentifyBehavior = import_IdentifyBehavior.IdentifyBehavior;
  OnOffSensorRequirements2.OnOffBehavior = import_OnOffBehavior.OnOffBehavior;
  OnOffSensorRequirements2.GroupsBehavior = import_GroupsBehavior.GroupsBehavior;
  OnOffSensorRequirements2.ScenesManagementBehavior = import_ScenesManagementBehavior.ScenesManagementBehavior;
  OnOffSensorRequirements2.LevelControlBehavior = import_LevelControlBehavior.LevelControlBehavior;
  OnOffSensorRequirements2.ColorControlBehavior = import_ColorControlBehavior.ColorControlBehavior;
  OnOffSensorRequirements2.server = { mandatory: { Identify: OnOffSensorRequirements2.IdentifyServer } };
  OnOffSensorRequirements2.client = {
    mandatory: { Identify: OnOffSensorRequirements2.IdentifyBehavior, OnOff: OnOffSensorRequirements2.OnOffBehavior },
    optional: {
      Groups: OnOffSensorRequirements2.GroupsBehavior,
      ScenesManagement: OnOffSensorRequirements2.ScenesManagementBehavior,
      LevelControl: OnOffSensorRequirements2.LevelControlBehavior,
      ColorControl: OnOffSensorRequirements2.ColorControlBehavior
    }
  };
})(OnOffSensorRequirements || (OnOffSensorRequirements = {}));
const OnOffSensorDeviceDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "OnOffSensor",
  deviceType: 2128,
  deviceRevision: 3,
  requirements: OnOffSensorRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)(OnOffSensorRequirements.server.mandatory.Identify)
});
const OnOffSensorDevice = OnOffSensorDeviceDefinition;
