// @project-chip/matter.js/dist/cjs/endpoint/definitions/device/ThermostatDevice.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ThermostatDevice_exports = {};
__export(ThermostatDevice_exports, {
  ThermostatDevice: () => ThermostatDevice,
  ThermostatDeviceDefinition: () => ThermostatDeviceDefinition,
  ThermostatRequirements: () => ThermostatRequirements
});
module.exports = __toCommonJS(ThermostatDevice_exports);
var import_IdentifyServer = require("../../../behavior/definitions/identify/IdentifyServer.js");
var import_ThermostatServer = require("../../../behavior/definitions/thermostat/ThermostatServer.js");
var import_GroupsServer = require("../../../behavior/definitions/groups/GroupsServer.js");
var import_ScenesManagementServer = require("../../../behavior/definitions/scenes-management/ScenesManagementServer.js");
var import_ThermostatUserInterfaceConfigurationServer = require("../../../behavior/definitions/thermostat-user-interface-configuration/ThermostatUserInterfaceConfigurationServer.js");
var import_EnergyPreferenceServer = require("../../../behavior/definitions/energy-preference/EnergyPreferenceServer.js");
var import_TimeSynchronizationServer = require("../../../behavior/definitions/time-synchronization/TimeSynchronizationServer.js");
var import_RelativeHumidityMeasurementBehavior = require("../../../behavior/definitions/relative-humidity-measurement/RelativeHumidityMeasurementBehavior.js");
var import_TimeSynchronizationBehavior = require("../../../behavior/definitions/time-synchronization/TimeSynchronizationBehavior.js");
var import_FanControlBehavior = require("../../../behavior/definitions/fan-control/FanControlBehavior.js");
var import_TemperatureMeasurementBehavior = require("../../../behavior/definitions/temperature-measurement/TemperatureMeasurementBehavior.js");
var import_OccupancySensingBehavior = require("../../../behavior/definitions/occupancy-sensing/OccupancySensingBehavior.js");
var import_MutableEndpoint = require("../../type/MutableEndpoint.js");
var import_SupportedBehaviors = require("../../properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var ThermostatRequirements;
((ThermostatRequirements2) => {
  ThermostatRequirements2.IdentifyServer = import_IdentifyServer.IdentifyServer;
  ThermostatRequirements2.ThermostatServer = import_ThermostatServer.ThermostatServer;
  ThermostatRequirements2.GroupsServer = import_GroupsServer.GroupsServer;
  ThermostatRequirements2.ScenesManagementServer = import_ScenesManagementServer.ScenesManagementServer;
  ThermostatRequirements2.ThermostatUserInterfaceConfigurationServer = import_ThermostatUserInterfaceConfigurationServer.ThermostatUserInterfaceConfigurationServer;
  ThermostatRequirements2.EnergyPreferenceServer = import_EnergyPreferenceServer.EnergyPreferenceServer;
  ThermostatRequirements2.TimeSynchronizationServer = import_TimeSynchronizationServer.TimeSynchronizationServer;
  ThermostatRequirements2.RelativeHumidityMeasurementBehavior = import_RelativeHumidityMeasurementBehavior.RelativeHumidityMeasurementBehavior;
  ThermostatRequirements2.TimeSynchronizationBehavior = import_TimeSynchronizationBehavior.TimeSynchronizationBehavior;
  ThermostatRequirements2.FanControlBehavior = import_FanControlBehavior.FanControlBehavior;
  ThermostatRequirements2.TemperatureMeasurementBehavior = import_TemperatureMeasurementBehavior.TemperatureMeasurementBehavior;
  ThermostatRequirements2.OccupancySensingBehavior = import_OccupancySensingBehavior.OccupancySensingBehavior;
  ThermostatRequirements2.server = {
    mandatory: { Identify: ThermostatRequirements2.IdentifyServer, Thermostat: ThermostatRequirements2.ThermostatServer },
    optional: {
      Groups: ThermostatRequirements2.GroupsServer,
      ScenesManagement: ThermostatRequirements2.ScenesManagementServer,
      ThermostatUserInterfaceConfiguration: ThermostatRequirements2.ThermostatUserInterfaceConfigurationServer,
      EnergyPreference: ThermostatRequirements2.EnergyPreferenceServer,
      TimeSynchronization: ThermostatRequirements2.TimeSynchronizationServer
    }
  };
  ThermostatRequirements2.client = {
    optional: {
      RelativeHumidityMeasurement: ThermostatRequirements2.RelativeHumidityMeasurementBehavior,
      TimeSynchronization: ThermostatRequirements2.TimeSynchronizationBehavior,
      FanControl: ThermostatRequirements2.FanControlBehavior,
      TemperatureMeasurement: ThermostatRequirements2.TemperatureMeasurementBehavior,
      OccupancySensing: ThermostatRequirements2.OccupancySensingBehavior
    },
    mandatory: {}
  };
})(ThermostatRequirements || (ThermostatRequirements = {}));
const ThermostatDeviceDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "Thermostat",
  deviceType: 769,
  deviceRevision: 3,
  requirements: ThermostatRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)(ThermostatRequirements.server.mandatory.Identify)
});
const ThermostatDevice = ThermostatDeviceDefinition;
