// @project-chip/matter.js/dist/cjs/endpoint/definitions/device/WaterLeakDetectorDevice.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var WaterLeakDetectorDevice_exports = {};
__export(WaterLeakDetectorDevice_exports, {
  WaterLeakDetectorDevice: () => WaterLeakDetectorDevice,
  WaterLeakDetectorDeviceDefinition: () => WaterLeakDetectorDeviceDefinition,
  WaterLeakDetectorRequirements: () => WaterLeakDetectorRequirements
});
module.exports = __toCommonJS(WaterLeakDetectorDevice_exports);
var import_IdentifyServer = require("../../../behavior/definitions/identify/IdentifyServer.js");
var import_BooleanStateServer = require("../../../behavior/definitions/boolean-state/BooleanStateServer.js");
var import_BooleanStateConfigurationServer = require("../../../behavior/definitions/boolean-state-configuration/BooleanStateConfigurationServer.js");
var import_MutableEndpoint = require("../../type/MutableEndpoint.js");
var import_SupportedBehaviors = require("../../properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var WaterLeakDetectorRequirements;
((WaterLeakDetectorRequirements2) => {
  WaterLeakDetectorRequirements2.IdentifyServer = import_IdentifyServer.IdentifyServer;
  WaterLeakDetectorRequirements2.BooleanStateServer = import_BooleanStateServer.BooleanStateServer.alter({ events: { stateChange: { optional: false } } });
  WaterLeakDetectorRequirements2.BooleanStateConfigurationServer = import_BooleanStateConfigurationServer.BooleanStateConfigurationServer;
  WaterLeakDetectorRequirements2.server = {
    mandatory: { Identify: WaterLeakDetectorRequirements2.IdentifyServer, BooleanState: WaterLeakDetectorRequirements2.BooleanStateServer },
    optional: { BooleanStateConfiguration: WaterLeakDetectorRequirements2.BooleanStateConfigurationServer }
  };
})(WaterLeakDetectorRequirements || (WaterLeakDetectorRequirements = {}));
const WaterLeakDetectorDeviceDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "WaterLeakDetector",
  deviceType: 67,
  deviceRevision: 1,
  requirements: WaterLeakDetectorRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)(
    WaterLeakDetectorRequirements.server.mandatory.Identify,
    WaterLeakDetectorRequirements.server.mandatory.BooleanState
  )
});
const WaterLeakDetectorDevice = WaterLeakDetectorDeviceDefinition;
