// @project-chip/matter.js/dist/cjs/endpoint/definitions/device/WindowCoveringControllerDevice.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var WindowCoveringControllerDevice_exports = {};
__export(WindowCoveringControllerDevice_exports, {
  WindowCoveringControllerDevice: () => WindowCoveringControllerDevice,
  WindowCoveringControllerDeviceDefinition: () => WindowCoveringControllerDeviceDefinition,
  WindowCoveringControllerRequirements: () => WindowCoveringControllerRequirements
});
module.exports = __toCommonJS(WindowCoveringControllerDevice_exports);
var import_IdentifyServer = require("../../../behavior/definitions/identify/IdentifyServer.js");
var import_WindowCoveringBehavior = require("../../../behavior/definitions/window-covering/WindowCoveringBehavior.js");
var import_IdentifyBehavior = require("../../../behavior/definitions/identify/IdentifyBehavior.js");
var import_GroupsBehavior = require("../../../behavior/definitions/groups/GroupsBehavior.js");
var import_ScenesManagementBehavior = require("../../../behavior/definitions/scenes-management/ScenesManagementBehavior.js");
var import_MutableEndpoint = require("../../type/MutableEndpoint.js");
var import_SupportedBehaviors = require("../../properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var WindowCoveringControllerRequirements;
((WindowCoveringControllerRequirements2) => {
  WindowCoveringControllerRequirements2.IdentifyServer = import_IdentifyServer.IdentifyServer;
  WindowCoveringControllerRequirements2.WindowCoveringBehavior = import_WindowCoveringBehavior.WindowCoveringBehavior;
  WindowCoveringControllerRequirements2.IdentifyBehavior = import_IdentifyBehavior.IdentifyBehavior;
  WindowCoveringControllerRequirements2.GroupsBehavior = import_GroupsBehavior.GroupsBehavior;
  WindowCoveringControllerRequirements2.ScenesManagementBehavior = import_ScenesManagementBehavior.ScenesManagementBehavior;
  WindowCoveringControllerRequirements2.server = { optional: { Identify: WindowCoveringControllerRequirements2.IdentifyServer }, mandatory: {} };
  WindowCoveringControllerRequirements2.client = {
    mandatory: { WindowCovering: WindowCoveringControllerRequirements2.WindowCoveringBehavior },
    optional: { Identify: WindowCoveringControllerRequirements2.IdentifyBehavior, Groups: WindowCoveringControllerRequirements2.GroupsBehavior, ScenesManagement: WindowCoveringControllerRequirements2.ScenesManagementBehavior }
  };
})(WindowCoveringControllerRequirements || (WindowCoveringControllerRequirements = {}));
const WindowCoveringControllerDeviceDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "WindowCoveringController",
  deviceType: 515,
  deviceRevision: 3,
  requirements: WindowCoveringControllerRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)()
});
const WindowCoveringControllerDevice = WindowCoveringControllerDeviceDefinition;
