// @project-chip/matter.js/dist/esm/protocol/interaction/InteractionMessenger.js
/**
 * @license
 * Copyright 2022-2024 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Status } from "../../cluster/globals/index.js";
import { SessionType } from "../../codec/MessageCodec.js";
import { ImplementationError, MatterFlowError, UnexpectedDataError } from "../../common/MatterError.js";
import { Logger } from "../../log/Logger.js";
import {
  RetransmissionLimitReachedError,
  UnexpectedMessageError
} from "../../protocol/MessageExchange.js";
import { TlvAny } from "../../tlv/TlvAny.js";
import {
  canAttributePayloadBeChunked,
  chunkAttributePayload,
  encodeAttributePayload,
  encodeEventPayload
} from "./AttributeDataEncoder.js";
import {
  TlvDataReport,
  TlvDataReportForSend,
  TlvInvokeRequest,
  TlvInvokeResponse,
  TlvReadRequest,
  TlvStatusResponse,
  TlvSubscribeRequest,
  TlvSubscribeResponse,
  TlvTimedRequest,
  TlvWriteRequest,
  TlvWriteResponse
} from "./InteractionProtocol.js";
import { INTERACTION_MODEL_REVISION } from "./InteractionServer.js";
import { StatusCode, StatusResponseError } from "./StatusCode.js";
var MessageType = /* @__PURE__ */ ((MessageType2) => {
  MessageType2[MessageType2["StatusResponse"] = 1] = "StatusResponse";
  MessageType2[MessageType2["ReadRequest"] = 2] = "ReadRequest";
  MessageType2[MessageType2["SubscribeRequest"] = 3] = "SubscribeRequest";
  MessageType2[MessageType2["SubscribeResponse"] = 4] = "SubscribeResponse";
  MessageType2[MessageType2["ReportData"] = 5] = "ReportData";
  MessageType2[MessageType2["WriteRequest"] = 6] = "WriteRequest";
  MessageType2[MessageType2["WriteResponse"] = 7] = "WriteResponse";
  MessageType2[MessageType2["InvokeRequest"] = 8] = "InvokeRequest";
  MessageType2[MessageType2["InvokeResponse"] = 9] = "InvokeResponse";
  MessageType2[MessageType2["TimedRequest"] = 10] = "TimedRequest";
  return MessageType2;
})(MessageType || {});
const logger = Logger.get("InteractionMessenger");
class InteractionMessenger {
  constructor(exchange) {
    this.exchange = exchange;
  }
  send(messageType, payload, options) {
    return this.exchange.send(messageType, payload, options);
  }
  sendStatus(status) {
    return this.send(
      1 /* StatusResponse */,
      TlvStatusResponse.encode({ status, interactionModelRevision: INTERACTION_MODEL_REVISION })
    );
  }
  async waitForSuccess() {
    await this.nextMessage(1 /* StatusResponse */);
  }
  async nextMessage(expectedMessageType) {
    const message = await this.exchange.nextMessage();
    const messageType = message.payloadHeader.messageType;
    this.throwIfErrorStatusMessage(message);
    if (expectedMessageType !== void 0 && messageType !== expectedMessageType) {
      throw new UnexpectedDataError(
        `Received unexpected message type: ${messageType}, expected: ${expectedMessageType}`
      );
    }
    return message;
  }
  async close() {
    await this.exchange.close();
  }
  throwIfErrorStatusMessage(message) {
    const {
      payloadHeader: { messageType },
      payload
    } = message;
    if (messageType !== 1 /* StatusResponse */) return;
    const { status } = TlvStatusResponse.decode(payload);
    if (status !== StatusCode.Success) throw new StatusResponseError(`Received error status: ${status}`, status);
  }
  getExchangeChannelName() {
    return this.exchange.channel.name;
  }
}
class InteractionServerMessenger extends InteractionMessenger {
  async handleRequest(recipient) {
    let continueExchange = true;
    let isGroupSession = false;
    try {
      while (continueExchange) {
        const message = await this.exchange.nextMessage();
        isGroupSession = message.packetHeader.sessionType === SessionType.Group;
        continueExchange = false;
        switch (message.payloadHeader.messageType) {
          case 2 /* ReadRequest */: {
            if (isGroupSession) {
              throw new StatusResponseError(
                `ReadRequest is not supported in group sessions`,
                Status.InvalidAction
              );
            }
            const readRequest = TlvReadRequest.decode(message.payload);
            await this.sendDataReport(
              await recipient.handleReadRequest(this.exchange, readRequest, message),
              readRequest.isFabricFiltered
            );
            break;
          }
          case 6 /* WriteRequest */: {
            const writeRequest = TlvWriteRequest.decode(message.payload);
            const { suppressResponse } = writeRequest;
            const writeResponse = await recipient.handleWriteRequest(this.exchange, writeRequest, message);
            if (!suppressResponse && !isGroupSession) {
              await this.send(7 /* WriteResponse */, TlvWriteResponse.encode(writeResponse));
            }
            break;
          }
          case 3 /* SubscribeRequest */: {
            const subscribeRequest = TlvSubscribeRequest.decode(message.payload);
            await recipient.handleSubscribeRequest(this.exchange, subscribeRequest, this, message);
            break;
          }
          case 8 /* InvokeRequest */: {
            const invokeRequest = TlvInvokeRequest.decode(message.payload);
            await recipient.handleInvokeRequest(this.exchange, invokeRequest, this, message);
            break;
          }
          case 10 /* TimedRequest */: {
            const timedRequest = TlvTimedRequest.decode(message.payload);
            recipient.handleTimedRequest(this.exchange, timedRequest, message);
            await this.sendStatus(StatusCode.Success);
            continueExchange = true;
            break;
          }
          default:
            throw new StatusResponseError(
              `Unsupported message type ${message.payloadHeader.messageType}`,
              Status.InvalidAction
            );
        }
      }
    } catch (error) {
      let errorStatusCode = StatusCode.Failure;
      if (error instanceof StatusResponseError) {
        logger.info(`Sending status response ${error.code} for interaction error: ${error.message}`);
        errorStatusCode = error.code;
      } else {
        logger.error(error);
      }
      if (!isGroupSession) {
        await this.sendStatus(errorStatusCode);
      }
    } finally {
      await this.exchange.close();
    }
  }
  /**
   * Handle DataReportPayload with the content of a DataReport to send, split them into multiple DataReport
   * messages and send them out based on the size.
   */
  async sendDataReport(dataReportPayload, forFabricFilteredRead) {
    const {
      subscriptionId,
      attributeReportsPayload,
      eventReportsPayload,
      suppressResponse,
      interactionModelRevision
    } = dataReportPayload;
    const dataReport = {
      subscriptionId,
      suppressResponse,
      interactionModelRevision,
      attributeReports: void 0,
      eventReports: void 0
    };
    if (attributeReportsPayload !== void 0 || eventReportsPayload !== void 0) {
      const attributeReportsToSend = [...attributeReportsPayload ?? []];
      const eventReportsToSend = [...eventReportsPayload ?? []];
      dataReport.moreChunkedMessages = true;
      const emptyDataReportBytes = TlvDataReportForSend.encode(dataReport);
      let firstAttributeAddedToReportMessage = false;
      let firstEventAddedToReportMessage = false;
      const sendAndResetReport = async () => {
        await this.sendDataReportMessage(dataReport);
        dataReport.attributeReports = void 0;
        dataReport.eventReports = void 0;
        messageSize = emptyDataReportBytes.length;
        firstAttributeAddedToReportMessage = false;
        firstEventAddedToReportMessage = false;
      };
      let messageSize = emptyDataReportBytes.length;
      while (true) {
        if (attributeReportsToSend.length > 0) {
          const attributeReport = attributeReportsToSend.shift();
          if (attributeReport !== void 0) {
            if (!firstAttributeAddedToReportMessage) {
              firstAttributeAddedToReportMessage = true;
              messageSize += 3;
            }
            const allowMissingFieldsForNonFabricFilteredRead = !forFabricFilteredRead && attributeReport.hasFabricSensitiveData;
            const encodedAttribute = encodeAttributePayload(attributeReport, {
              allowMissingFieldsForNonFabricFilteredRead
            });
            const attributeReportBytes = TlvAny.getEncodedByteLength(encodedAttribute);
            if (messageSize + attributeReportBytes > this.exchange.maxPayloadSize) {
              if (canAttributePayloadBeChunked(attributeReport)) {
                attributeReportsToSend.unshift(...chunkAttributePayload(attributeReport));
                continue;
              }
              await sendAndResetReport();
            }
            messageSize += attributeReportBytes;
            if (dataReport.attributeReports === void 0) dataReport.attributeReports = [];
            dataReport.attributeReports.push(encodedAttribute);
          }
        } else if (eventReportsToSend.length > 0) {
          const eventReport = eventReportsToSend.shift();
          if (eventReport === void 0) {
            delete dataReport.moreChunkedMessages;
            break;
          }
          if (!firstEventAddedToReportMessage) {
            firstEventAddedToReportMessage = true;
            messageSize += 3;
          }
          const allowMissingFieldsForNonFabricFilteredRead = !forFabricFilteredRead && eventReport.hasFabricSensitiveData;
          const encodedEvent = encodeEventPayload(eventReport, {
            allowMissingFieldsForNonFabricFilteredRead
          });
          const eventReportBytes = TlvAny.getEncodedByteLength(encodedEvent);
          if (messageSize + eventReportBytes > this.exchange.maxPayloadSize) {
            await sendAndResetReport();
          }
          messageSize += eventReportBytes;
          if (dataReport.eventReports === void 0) dataReport.eventReports = [];
          dataReport.eventReports.push(encodedEvent);
        } else {
          delete dataReport.moreChunkedMessages;
          break;
        }
      }
    }
    await this.sendDataReportMessage(dataReport);
  }
  async sendDataReportMessage(dataReport) {
    const dataReportToSend = {
      ...dataReport,
      suppressResponse: dataReport.moreChunkedMessages ? false : dataReport.suppressResponse
      // always false when moreChunkedMessages is true
    };
    const encodedMessage = TlvDataReportForSend.encode(dataReportToSend);
    if (encodedMessage.length > this.exchange.maxPayloadSize) {
      throw new MatterFlowError(
        `DataReport is too long to fit in a single chunk, This should not happen! Data: ${Logger.toJSON(
          dataReportToSend
        )}`
      );
    }
    logger.debug(
      `Sending DataReport chunk with ${dataReportToSend.attributeReports?.length ?? 0} attributes and ${dataReportToSend.eventReports?.length ?? 0} events: ${encodedMessage.length} bytes (moreChunkedMessages: ${dataReportToSend.moreChunkedMessages ?? false}, suppressResponse: ${dataReportToSend.suppressResponse})`
    );
    if (dataReportToSend.suppressResponse) {
      try {
        await this.exchange.send(5 /* ReportData */, encodedMessage, {
          expectAckOnly: true
        });
      } catch (e) {
        UnexpectedMessageError.accept(e);
        const { receivedMessage } = e;
        this.throwIfErrorStatusMessage(receivedMessage);
      }
    } else {
      await this.exchange.send(5 /* ReportData */, encodedMessage);
      await this.waitForSuccess();
    }
  }
}
class IncomingInteractionClientMessenger extends InteractionMessenger {
  async readDataReports(sendFinalAckMessage = true) {
    let subscriptionId;
    const attributeValues = [];
    const eventValues = [];
    while (true) {
      const dataReportMessage = await this.exchange.waitFor(5 /* ReportData */);
      const report = TlvDataReport.decode(dataReportMessage.payload);
      if (subscriptionId === void 0 && report.subscriptionId !== void 0) {
        subscriptionId = report.subscriptionId;
      } else if ((subscriptionId !== void 0 || report.subscriptionId !== void 0) && report.subscriptionId !== subscriptionId) {
        throw new UnexpectedDataError(`Invalid subscription ID ${report.subscriptionId} received`);
      }
      logger.debug(
        `Received DataReport chunk with ${report.attributeReports?.length ?? 0} attributes and ${report.eventReports?.length ?? 0} events, suppressResponse: ${report.suppressResponse}, moreChunkedMessages: ${report.moreChunkedMessages}`
      );
      if (Array.isArray(report.attributeReports) && report.attributeReports.length > 0) {
        attributeValues.push(...report.attributeReports);
      }
      if (Array.isArray(report.eventReports) && report.eventReports.length > 0) {
        eventValues.push(...report.eventReports);
      }
      if (report.moreChunkedMessages || sendFinalAckMessage && !report.suppressResponse) {
        await this.sendStatus(StatusCode.Success);
      }
      if (!report.moreChunkedMessages) {
        report.attributeReports = attributeValues;
        report.eventReports = eventValues;
        return report;
      }
    }
  }
}
class InteractionClientMessenger extends IncomingInteractionClientMessenger {
  constructor(exchangeProvider) {
    super(exchangeProvider.initiateExchange());
    this.exchangeProvider = exchangeProvider;
  }
  /** Implements a send method with an automatic reconnection mechanism */
  async send(messageType, payload, options) {
    if (this.exchange.channel.closed) {
      throw new ImplementationError("The exchange channel is closed. Please connect the device first.");
    }
    try {
      return await this.exchange.send(messageType, payload, options);
    } catch (error) {
      if (error instanceof RetransmissionLimitReachedError) {
        await this.exchange.close();
        if (await this.exchangeProvider.reconnectChannel()) {
          this.exchange = this.exchangeProvider.initiateExchange();
          return await this.exchange.send(messageType, payload);
        }
      }
      throw error;
    }
  }
  async sendReadRequest(readRequest) {
    await this.send(2 /* ReadRequest */, TlvReadRequest.encode(readRequest));
    return this.readDataReports();
  }
  async sendSubscribeRequest(subscribeRequest) {
    await this.send(3 /* SubscribeRequest */, TlvSubscribeRequest.encode(subscribeRequest));
    const report = await this.readDataReports();
    const { subscriptionId } = report;
    if (subscriptionId === void 0) {
      throw new UnexpectedDataError(`Subscription ID not provided in report`);
    }
    const subscribeResponseMessage = await this.nextMessage(4 /* SubscribeResponse */);
    const subscribeResponse = TlvSubscribeResponse.decode(subscribeResponseMessage.payload);
    if (subscribeResponse.subscriptionId !== subscriptionId) {
      throw new MatterFlowError(
        `Received subscription ID ${subscribeResponse.subscriptionId} instead of ${subscriptionId}`
      );
    }
    return {
      subscribeResponse,
      report
    };
  }
  async sendInvokeCommand(invokeRequest, minimumResponseTimeoutMs) {
    if (invokeRequest.suppressResponse) {
      await this.requestWithSuppressedResponse(
        8 /* InvokeRequest */,
        TlvInvokeRequest,
        invokeRequest,
        minimumResponseTimeoutMs
      );
    } else {
      return await this.request(
        8 /* InvokeRequest */,
        TlvInvokeRequest,
        9 /* InvokeResponse */,
        TlvInvokeResponse,
        invokeRequest,
        minimumResponseTimeoutMs
      );
    }
  }
  async sendWriteCommand(writeRequest) {
    if (writeRequest.suppressResponse) {
      await this.requestWithSuppressedResponse(6 /* WriteRequest */, TlvWriteRequest, writeRequest);
    } else {
      return await this.request(
        6 /* WriteRequest */,
        TlvWriteRequest,
        7 /* WriteResponse */,
        TlvWriteResponse,
        writeRequest
      );
    }
  }
  sendTimedRequest(timeoutSeconds) {
    return this.request(10 /* TimedRequest */, TlvTimedRequest, 1 /* StatusResponse */, TlvStatusResponse, {
      timeout: timeoutSeconds,
      interactionModelRevision: INTERACTION_MODEL_REVISION
    });
  }
  async requestWithSuppressedResponse(requestMessageType, requestSchema, request, minimumResponseTimeoutMs) {
    await this.send(requestMessageType, requestSchema.encode(request), {
      expectAckOnly: true,
      minimumResponseTimeoutMs
    });
  }
  async request(requestMessageType, requestSchema, responseMessageType, responseSchema, request, minimumResponseTimeoutMs) {
    await this.send(requestMessageType, requestSchema.encode(request), {
      expectAckOnly: false,
      minimumResponseTimeoutMs
    });
    const responseMessage = await this.nextMessage(responseMessageType);
    return responseSchema.decode(responseMessage.payload);
  }
}
export {
  IncomingInteractionClientMessenger,
  InteractionClientMessenger,
  InteractionServerMessenger,
  MessageType
};
